{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f8728aad-76a1-44eb-8f3e-641d9d0d2f8d",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "78eb35ab-e3c0-40e7-9827-8c9f88fb9976",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "from Bio import AlignIO\n",
    "import sys\n",
    "import PyQt5\n",
    "from ete3 import Tree, TreeStyle, TextFace, NodeStyle, AttrFace\n",
    "import math\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# my functions\n",
    "sys.path.append(os.path.join('..', 'scripts', 'python'))\n",
    "from create_phyl_tree import run_mafft, alignment_stats, run_raxmlng_check, run_raxmlng_tree, run_raxmlng_bootstrap, run_raxmlng_support, filter_fasta_file, shorten_sequence_names, num_seqs, run_gblocks_grid_search, move_search_grid_files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69035ef1-4337-431f-a71a-1ada3deed9d4",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "209c4951-3a50-488a-8868-1ec57a17d857",
   "metadata": {},
   "outputs": [],
   "source": [
    "project = 'Suthaus_2022'\n",
    "marker = 'Full18S'\n",
    "sim = 'sim_90'\n",
    "denoise_method = 'RAD'\n",
    "raw_data = os.path.join('..', 'raw_data')\n",
    "results_path = os.path.join('..', 'results')\n",
    "reference_align_dir = os.path.join(raw_data, 'reference_alignments', 'vamp')\n",
    "new_seq_dir = os.path.join(results_path, 'tax_assignment_vsearch', project, marker, sim, denoise_method, 'vamp_specific')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9111e557-36e4-4b71-ae0a-28aa49ccd064",
   "metadata": {},
   "source": [
    "# Preparing input fasta file for the sequence alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cbeb006-9075-4592-b77c-48760e5f093d",
   "metadata": {},
   "source": [
    "### Create a fasta file with all the new sequences assigned to the vampyrellids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3afd623-1782-4452-8040-634306929824",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to the output file\n",
    "merged_fasta_file = os.path.join(new_seq_dir, 'all_samples.fasta')\n",
    "\n",
    "# Create a list of fasta files, excluding those starting with 'Mock'\n",
    "files = [file for file in os.listdir(new_seq_dir) if not file.startswith('Mock')]\n",
    "\n",
    "# List to store the modified lines\n",
    "merged_lines = []\n",
    "\n",
    "for file in files:\n",
    "    path = os.path.join(new_seq_dir, file)\n",
    "    sample = file.split('_')[0]\n",
    "    with open(path, 'r') as infile:\n",
    "        for line in infile:\n",
    "            if line.startswith('>'):\n",
    "                # Replace ';' with '_' in the line as the RAxML-ng does not support the semicolon\n",
    "                modified_line = line.strip().replace(';', '_')\n",
    "                # Add the sample name to the line\n",
    "                merged_lines.append(modified_line + '_' + sample + '\\n')\n",
    "            else:\n",
    "                merged_lines.append(line)\n",
    "\n",
    "# Write the contents to the output file: results -> tax_assignment_vsearch -> vamp_specific -> all_samples.fasta\n",
    "with open(merged_fasta_file, 'w') as outfile:\n",
    "    outfile.writelines(merged_lines)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d8591c7-c6d8-4cae-b8b5-227e42b5e312",
   "metadata": {},
   "source": [
    "### Prepare vampyrella reference sequence database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c352511-a31b-4fdd-bbee-a3569bfb26ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We have to exclude the unpublished sequences\n",
    "# All the unpublished sequences started with 'XXXXXXXX' in the sequence ID\n",
    "\n",
    "# Vampyrella reference sequences path\n",
    "vamp_ref_seqs = os.path.join(reference_align_dir, '2023_VAMPYRELLIDA_SSU_annotated_PR2_AM_AS_not_aligned_version.fasta')\n",
    "# Vampyrella filtered fasta file\n",
    "output_dir = os.path.join(results_path, 'phyl_trees', project, marker, denoise_method, sim)\n",
    "filtered_fasta_file = os.path.join(output_dir, 'vamp_ref_seqs.fasta')\n",
    "\n",
    "# Create path to the save directory if not exist\n",
    "if not os.path.exists(output_dir):\n",
    "    os.makedirs(output_dir)\n",
    "\n",
    "# Filter out sequences with headers starting with \"XXXXXXXX\", and return the filtered lines.\n",
    "filtered_lines = filter_fasta_file(file_path = vamp_ref_seqs, exclude_pattern = 'XXXXXXXX')\n",
    "shorten_seq_names = shorten_sequence_names(filtered_lines)\n",
    "\n",
    "# Write the contents to the output fasta file file: results -> phyl_trees -> vamp_ref_seqs.fasta\n",
    "with open(filtered_fasta_file, 'w') as outfile:\n",
    "    outfile.writelines(shorten_seq_names)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6faf1b7-9c49-4b2b-90b4-a147bdeb7e94",
   "metadata": {},
   "source": [
    "### Combine the new sequences with our vampyrella reference sequence database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dbb23848-137f-4d88-8b2b-990706e29b70",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Combined files ../results/phyl_trees/all_projects_combined/extracted_18S_merged.fasta and ../results/phyl_trees/Suthaus_2022/Full18S/RAD/sim_90/vamp_ref_seqs.fasta into ../results/phyl_trees/all_projects_combined/vamp_ref_and_new_seqs_comb.fasta\n"
     ]
    }
   ],
   "source": [
    "# Reference alignment path\n",
    "vamp_ref_seqs = os.path.join(results_path, 'phyl_trees', project, marker, denoise_method, sim, 'vamp_ref_seqs.fasta')\n",
    "\n",
    "# Path to the new sequences\n",
    "# new_vamp_seqs = os.path.join(new_seq_dir, 'all_samples.fasta')\n",
    "\n",
    "new_vamp_seqs = os.path.join(results_path, 'phyl_trees', 'all_projects_combined', 'extracted_18S_merged.fasta')\n",
    "\n",
    "# Path to the output directory and name of the output file\n",
    "# output_dir = os.path.join(results_path, 'phyl_trees', project, marker, denoise_method, sim)\n",
    "output_dir = os.path.join(results_path, 'phyl_trees', 'all_projects_combined')\n",
    "\n",
    "combined_alignment = os.path.join(output_dir, 'vamp_ref_and_new_seqs_comb.fasta')\n",
    "\n",
    "# Combine ref. sequences fasta file and new vamp sequences\n",
    "\n",
    "# Read the contents of the reference sequences file\n",
    "with open(vamp_ref_seqs, 'r') as file1:\n",
    "    content1 = file1.read()\n",
    "\n",
    "# Read the contents of the denoised sequences file\n",
    "with open(new_vamp_seqs, 'r') as file2:\n",
    "    content2 = file2.read()\n",
    "\n",
    "# Write the combined contents to the new file: results -> phyl_trees -> vamp_ref_and_new_seqs_comb.fasta\n",
    "with open(combined_alignment, 'w') as infile:\n",
    "    infile.write(content1 + '\\n' + content2)\n",
    "\n",
    "print(f\"Combined files {new_vamp_seqs} and {vamp_ref_seqs} into {combined_alignment}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f55fcd5e-deb8-4fef-9d93-eae424c097d3",
   "metadata": {},
   "source": [
    "# Create alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "33cb856f-f428-4ee8-a48f-df33abfe898a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment was saved to: ../results/phyl_trees/all_projects_combined/vamp_mafft.fasta\n",
      "Log file was saved to: ../results/phyl_trees/all_projects_combined/vamp_mafft.log\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Creating alignment\n",
    "# input_sequences = os.path.join(results_path, 'phyl_trees', project, marker, denoise_method, sim, 'vamp_ref_and_new_seqs_comb.fasta')\n",
    "input_sequences = os.path.join(results_path, 'phyl_trees', 'all_projects_combined', 'vamp_ref_and_new_seqs_comb.fasta')\n",
    "\n",
    "# output_sequences = os.path.join(results_path, 'phyl_trees', project, marker, denoise_method, sim, 'alignments', 'vamp_mafft.fasta')\n",
    "output_sequences = os.path.join(results_path, 'phyl_trees', 'all_projects_combined', 'vamp_mafft.fasta')\n",
    "\n",
    "# log_file = os.path.join(results_path, 'phyl_trees', project, marker, denoise_method, sim, 'alignments', 'vamp_mafft.log')\n",
    "log_file = os.path.join(results_path, 'phyl_trees', 'all_projects_combined', 'vamp_mafft.log')\n",
    "\n",
    "\n",
    "run_mafft(input_file = input_sequences, \n",
    "          output_file = output_sequences, \n",
    "          log_file = log_file, \n",
    "          method = 'localpair', \n",
    "          maxiterate_num = 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5cbc541-ea54-4e8b-bc03-0be1adfaeb3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Alignment stats\n",
    "alignment = os.path.join(results_path, 'phyl_trees', project, marker, denoise_method, sim, 'alignments', 'vamp_mafft.fasta')\n",
    "\n",
    "\n",
    "num_sequences, alignment_length, average_gaps, percentage_gaps, percentage_identity = alignment_stats(alignment_file = alignment)\n",
    "\n",
    "print(f'''\n",
    "Number of sequences: {num_sequences}\n",
    "Alignment length: {alignment_length} base pairs\n",
    "Average number of gaps per sequence: {round(average_gaps, 2)}\n",
    "Percentage gaps: {round(percentage_gaps, 2)}%\n",
    "Percentage identity: {round(percentage_identity, 2)}%\n",
    "''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cee194a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Alignment stats\n",
    "# alignment = os.path.join(results_path, 'phyl_trees', project, marker, denoise_method, sim, 'alignments', 'vamp_mafft.fasta')\n",
    "\n",
    "alignment = os.path.join(results_path, 'phyl_trees', 'refrence_alignment_tree', 'vamp_mafft.fasta')\n",
    "\n",
    "\n",
    "num_sequences, alignment_length, average_gaps, percentage_gaps, percentage_identity = alignment_stats(alignment_file = alignment)\n",
    "\n",
    "print(f'''\n",
    "Number of sequences: {num_sequences}\n",
    "Alignment length: {alignment_length} base pairs\n",
    "Average number of gaps per sequence: {round(average_gaps, 2)}\n",
    "Percentage gaps: {round(percentage_gaps, 2)}%\n",
    "Percentage identity: {round(percentage_identity, 2)}%\n",
    "''')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d285cad-6c83-4b1b-9899-40dc614fa1a0",
   "metadata": {},
   "source": [
    "# Check alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25d26c91-fd02-4620-8abb-0bac59713182",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the alignment manually to see how does it look like"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9294b09-a271-4276-9dc4-a65f1c987ff1",
   "metadata": {},
   "source": [
    "# Gblocks: masking poorly aligned positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a8420cd-b7bc-451d-a63f-f220246ba2df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the Parameter Range\n",
    "Defining the range of values we want to explore for each parameter\n",
    "mafft_alignment = os.path.join(results_path, 'phyl_trees', project, marker, denoise_method, sim, 'alignments', 'vamp_mafft.fasta')\n",
    "\n",
    "# Number of sequences and half of the sequences in the alignemnt as a baseline for Gblocks parameters\n",
    "num_seqs_ = num_seqs(mafft_alignment)\n",
    "print(f'Total number of sequences in the alignment: {num_seqs_}')\n",
    "half_num_seqs = math.floor(num_seqs_ / 2) + 1 # half of the sequences, rounded down\n",
    "print(f'Half of the sequences in the alignment (rounded down): {half_num_seqs}')\n",
    "\n",
    "# b1\n",
    "# Description: Minimum Number of Sequences for a Conserved Position\n",
    "# Minimum is half of the sequences in the alignment.\n",
    "step_size_b1 = 20\n",
    "start_b1 = 0\n",
    "stop_b1 = step_size_b1 * 10\n",
    "b1_values = [half_num_seqs + i for i in range(start_b1, stop_b1, step_size_b1)]\n",
    "\n",
    "# b2\n",
    "# Description: Minimum Number Of Sequences For A Flank Position.\n",
    "# The lower limit for b2 is the value you set for b1.\n",
    "step_size_b2 = 20\n",
    "start_b2 = 0\n",
    "stop_b2 = step_size_b2 * 10\n",
    "b2_values = [half_num_seqs + i for i in range(start_b2, stop_b2, step_size_b2)]\n",
    "\n",
    "# b3:\n",
    "# Description: Maximum Number Of Contiguous Nonconserved Positions.\n",
    "# Setting this value too high can include regions of the alignment that are too variable, \n",
    "# whereas setting it too low can exclude potentially informative parts of the alignment.\n",
    "# Default = 8.\n",
    "step_size_b3 = 5\n",
    "start_b3 = 8\n",
    "stop_b3 = (step_size_b3 * 10) + start_b3\n",
    "b3_values = list(range(start_b3, stop_b3, step_size_b3))\n",
    "\n",
    "# b4:\n",
    "# Description: Minimum Length Of A Block.\n",
    "# Any block that's shorter than the number set will be excluded from the final alignment.\n",
    "# Default = 10.\n",
    "step_size_b4 = 5\n",
    "start_b4 = 10\n",
    "stop_b4 =  (step_size_b4 * 10) + start_b4\n",
    "b4_values = list(range(start_b4, stop_b4, step_size_b4))\n",
    "\n",
    "# v5\n",
    "# Description: Allowed Gap Positions.\n",
    "# a: all gaps allowed (the most relaxed setting)\n",
    "# h: half (a balanced approach)\n",
    "# n: no gaps allowed (the most strict setting)\n",
    "b5_values = ['a', 'h', 'n']\n",
    "\n",
    "print(f'''\n",
    "b1: {b1_values}\n",
    "b2: {b2_values}\n",
    "b3: {b3_values}\n",
    "b4: {b4_values}\n",
    "b5: {b5_values}\n",
    "''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c46014a-4508-410a-aa63-247f4d7ff5c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "alignments_path = os.path.join(results_path, 'phyl_trees', project, marker, denoise_method, sim, 'alignments')\n",
    "alignment = os.path.join(alignments_path, 'vamp_mafft.fasta')\n",
    "gblocks_path = os.path.join('..', 'raw_data', 'packages', 'Gblocks_0.91b', 'Gblocks')\n",
    "\n",
    "\n",
    "run_gblocks_grid_search(mafft_alignment = alignment,\n",
    "                        b1_values = b1_values,\n",
    "                        b2_values = b2_values,\n",
    "                        b3_values = b3_values,\n",
    "                        b4_values = b4_values,\n",
    "                        b5_values = b5_values,\n",
    "                        gblocks_path = gblocks_path)\n",
    "\n",
    "# Move all the output files to the separate 'grid_search' directory\n",
    "move_search_grid_files(directory_path = alignments_path, \n",
    "                       grid_search_path = os.path.join(alignments_path, 'grid_search'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c3034c8-6e96-4fa7-82b3-b277f88ba8a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Checking some basic statistics about our MAFFT alignment\n",
    "alignments_path = os.path.join(results_path, 'phyl_trees', project, marker, denoise_method, sim, 'alignments')\n",
    "grid_search_path = os.path.join(alignments_path, 'grid_search')\n",
    "\n",
    "fasta_files = [fasta for fasta in os.listdir(grid_search_path) if '.fasta' in fasta and '_h_' in fasta]\n",
    "\n",
    "for fasta_file in fasta_files:\n",
    "    alignment_path = os.path.join(grid_search_path, fasta_file)\n",
    "    parameters = fasta_file.split('grid_')[1].strip('.fasta')\n",
    "    print(f'Paramters: {parameters}')\n",
    "    num_sequences, alignment_length, average_gaps, percentage_gaps, percentage_identity = alignment_stats(alignment_file = alignment_path)\n",
    "    print(f\"Number of sequences: {num_sequences}\")\n",
    "    print(f\"Alignment length: {alignment_length}\")\n",
    "    print(f\"Average number of gaps per sequence: {average_gaps:.2f}\")\n",
    "    print(f\"Percentage of gaps across the sequences: {percentage_gaps:.2f}%\")\n",
    "    print(f\"Average percentage identity: {percentage_identity:.2f}%\")\n",
    "    print('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3f9a44e-99a9-4fa4-b61c-cec033877560",
   "metadata": {},
   "outputs": [],
   "source": [
    "alignments_path = os.path.join(results_path, 'phyl_trees', project, marker, denoise_method, sim, 'alignments')\n",
    "grid_search_path = os.path.join(alignments_path, 'grid_search')\n",
    "fasta_files = [fasta for fasta in os.listdir(grid_search_path) if '.fasta' in fasta and '_h_' in fasta]\n",
    "results = []\n",
    "\n",
    "for fasta_file in fasta_files:\n",
    "    metrics = alignment_stats(os.path.join(grid_search_path, fasta_file))\n",
    "    b5_value = fasta_file.split(\"_grid_\")[1].split(\"_\")[0]\n",
    "    iteration_num = fasta_file.split(\"_grid_\")[1].split(\"_\")[1].strip('.fasta')\n",
    "    results.append((b5_value, iteration_num) + metrics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63f48e3d-ea17-4994-b426-a760adb86244",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a dataframe from the grid results\n",
    "grid_df = pd.DataFrame(results, columns=['b5_value', \n",
    "                                         'iteration_num', \n",
    "                                         'num_sequences', \n",
    "                                         'alignment_length', \n",
    "                                         'avg_num_gaps', \n",
    "                                         'percentage_gaps', \n",
    "                                         'avg_percentage_identity'])\n",
    "# Sort the dataframe based on the iteration number\n",
    "grid_df = grid_df.sort_values(by='iteration_num').reset_index(drop=True)\n",
    "\n",
    "# Add the values from the Gblocks parameters\n",
    "grid_df['b1_value'] = b1_values\n",
    "grid_df['b2_value'] = b2_values\n",
    "grid_df['b3_value'] = b3_values\n",
    "grid_df['b4_value'] = b4_values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0541fc3-eeb5-4a9f-ac50-92218884d5ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "grid_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bcc36f4-cb95-4c74-9cc5-4aa0032484dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "alignments_path = os.path.join(results_path, 'phyl_trees', project, marker, denoise_method, sim, 'alignments')\n",
    "grid_search_path = os.path.join(alignments_path, 'grid_search')\n",
    "\n",
    "# Visualization\n",
    "\n",
    "# Set style of seaborn plot\n",
    "sns.set(style=\"white\")\n",
    "\n",
    "# Create a 2x2 grid of subplots\n",
    "fig, axes = plt.subplots(2, 2, figsize=(14, 10))\n",
    "\n",
    "# Define the metrics and their respective labels\n",
    "metrics = ['alignment_length', 'avg_num_gaps', 'percentage_gaps', 'avg_percentage_identity']\n",
    "labels = ['Alignment Length', 'Average Number of Gaps', 'Percentage of Gaps', 'Average Percentage Identity']\n",
    "\n",
    "# Iterate over each metric and plot it in its subplot\n",
    "for ax, metric, label in zip(axes.ravel(), metrics, labels):\n",
    "    sns.lineplot(x='iteration_num', y=metric, data=grid_df, color='lightcoral', ax=ax)\n",
    "    ax.set_title(label + ' vs Iteration Number', fontweight='bold')\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_ylabel(label)\n",
    "    # Set visibility for spines\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['bottom'].set_visible(True)\n",
    "    ax.spines['left'].set_visible(True)\n",
    "\n",
    "# # Adjust layout for better spacing\n",
    "# plt.tight_layout()\n",
    "# plt.show()\n",
    "\n",
    "# save path\n",
    "save_path = os.path.join(grid_search_path, 'pr2_ref_align_gblocks_grid_search_h.png')\n",
    "# Save the figure\n",
    "# plt.savefig(save_path, dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1af09fe6-3951-4134-a8df-e8bd5f942065",
   "metadata": {},
   "source": [
    "# Create Maximum Likelihood tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "850d26c3-0acc-4ada-8b63-4ba986227a1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Variables:\n",
    "# alignment = os.path.join(results_path, 'phyl_trees', project, marker, denoise_method, sim, 'alignments', 'vamp_mafft.fasta')\n",
    "model = 'GTR+G'\n",
    "\n",
    "# alignment = os.path.join(results_path, 'phyl_trees', 'refrence_alignment_tree', 'vamp_mafft_cut.fasta')\n",
    "\n",
    "alignment = os.path.join(results_path, 'phyl_trees', 'all_projects_combined', 'vamp_mafft.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f1a53f7d-3908-4bbc-9108-500c373914aa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "RAxML-NG v. 1.1.0 released on 29.11.2021 by The Exelixis Lab.\n",
      "Developed by: Alexey M. Kozlov and Alexandros Stamatakis.\n",
      "Contributors: Diego Darriba, Tomas Flouri, Benoit Morel, Sarah Lutteropp, Ben Bettisworth.\n",
      "Latest version: https://github.com/amkozlov/raxml-ng\n",
      "Questions/problems/suggestions? Please visit: https://groups.google.com/forum/#!forum/raxml\n",
      "\n",
      "System: 11th Gen Intel(R) Core(TM) i7-11850H @ 2.50GHz, 8 cores, 11 GB RAM\n",
      "\n",
      "RAxML-NG was called at 11-Dec-2023 11:53:07 as follows:\n",
      "\n",
      "raxml-ng --check --msa ../results/phyl_trees/all_projects_combined/vamp_mafft.fasta --model GTR+G --prefix T1_data_check\n",
      "\n",
      "Analysis options:\n",
      "  run mode: Alignment validation\n",
      "  start tree(s): \n",
      "  random seed: 1702291987\n",
      "  SIMD kernels: AVX2\n",
      "  parallelization: coarse-grained (auto), PTHREADS (auto)\n",
      "\n",
      "[00:00:00] Reading alignment from file: ../results/phyl_trees/all_projects_combined/vamp_mafft.fasta\n",
      "[00:00:00] Loaded alignment with 183 taxa and 2038 sites\n",
      "\n",
      "WARNING: Fully undetermined columns found: 7\n",
      "\n",
      "WARNING: Sequences Thalassomyxa_strain_KibAr_KC779519 and Thalassomyxa_strain_V1ld4_KC779520 are exactly identical!\n",
      "WARNING: Sequences 93ec799fa70cc9a684fdbec04fb423d8_Sim17 and 93ec799fa70cc9a684fdbec04fb423d8_Sim22 are exactly identical!\n",
      "WARNING: Sequences 91b1f98492eba26f9f791cf7949afb56_NH4 and 91b1f98492eba26f9f791cf7949afb56_Sim17 are exactly identical!\n",
      "WARNING: Sequences 91b1f98492eba26f9f791cf7949afb56_NH4 and 91b1f98492eba26f9f791cf7949afb56_Sim22 are exactly identical!\n",
      "WARNING: Sequences 91b1f98492eba26f9f791cf7949afb56_NH4 and 91b1f98492eba26f9f791cf7949afb56_Th38 are exactly identical!\n",
      "WARNING: Sequences 91b1f98492eba26f9f791cf7949afb56_NH4 and 91b1f98492eba26f9f791cf7949afb56_Th40 are exactly identical!\n",
      "WARNING: Sequences fa04010ef186193e6af2567e8e77c9fd_Sim17 and fa04010ef186193e6af2567e8e77c9fd_Sim22 are exactly identical!\n",
      "WARNING: Sequences Placopus_X_GU385615 and Placopus_XX_GU385647 are exactly identical!\n",
      "WARNING: Duplicate sequences found: 8\n",
      "\n",
      "NOTE: Reduced alignment (with duplicates and gap-only sites/taxa removed) \n",
      "NOTE: was saved to: /home/lubo/code/wRajter/vampyrella_2023/notebooks/T1_data_check.raxml.reduced.phy\n",
      "\n",
      "Alignment comprises 1 partitions and 2031 sites\n",
      "\n",
      "Partition 0: noname\n",
      "Model: GTR+FO+G4m\n",
      "Alignment sites: 2031\n",
      "Gaps: 32.33 %\n",
      "Invariant sites: 41.36 %\n",
      "\n",
      "\n",
      "\n",
      "Alignment can be successfully read by RAxML-NG.\n",
      "\n",
      "\n",
      "Execution log saved to: /home/lubo/code/wRajter/vampyrella_2023/notebooks/T1_data_check.raxml.log\n",
      "\n",
      "Analysis started: 11-Dec-2023 11:53:07 / finished: 11-Dec-2023 11:53:07\n",
      "\n",
      "Elapsed time: 0.081 seconds\n",
      "\n",
      "\n",
      "Output files were moved to: ../results/phyl_trees/all_projects_combined/T1_data_check\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # Run raxml-ng (https://github.com/amkozlov/raxml-ng/wiki/Tutorial) to check the alignment and save the output in the specified directory.\n",
    "# prefix = 'T1'\n",
    "# output_dir = os.path.join(results_path, 'phyl_trees', project, marker, denoise_method, sim, prefix)\n",
    "\n",
    "prefix = 'T1_data_check'\n",
    "# output_dir = os.path.join(results_path, 'phyl_trees', 'refrence_alignment_tree', prefix)\n",
    "output_dir = os.path.join(results_path, 'phyl_trees', 'all_projects_combined', prefix)\n",
    "\n",
    "# Check if the output directory exists, create it if not\n",
    "if not os.path.exists(output_dir):\n",
    "    os.makedirs(output_dir)\n",
    "\n",
    "run_raxmlng_check(alignment = alignment, \n",
    "                  output_dir = output_dir, \n",
    "                  model = model, \n",
    "                  prefix = prefix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7c3acebd-b22a-4438-85bf-0663d22e1792",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "RAxML-NG v. 1.1.0 released on 29.11.2021 by The Exelixis Lab.\n",
      "Developed by: Alexey M. Kozlov and Alexandros Stamatakis.\n",
      "Contributors: Diego Darriba, Tomas Flouri, Benoit Morel, Sarah Lutteropp, Ben Bettisworth.\n",
      "Latest version: https://github.com/amkozlov/raxml-ng\n",
      "Questions/problems/suggestions? Please visit: https://groups.google.com/forum/#!forum/raxml\n",
      "\n",
      "System: 11th Gen Intel(R) Core(TM) i7-11850H @ 2.50GHz, 8 cores, 11 GB RAM\n",
      "\n",
      "RAxML-NG was called at 11-Dec-2023 11:53:33 as follows:\n",
      "\n",
      "raxml-ng --msa ../results/phyl_trees/all_projects_combined/vamp_mafft.fasta --model GTR+G --prefix T2_MLtree --threads 3 --seed 2 --tree pars{25},rand{25}\n",
      "\n",
      "Analysis options:\n",
      "  run mode: ML tree search\n",
      "  start tree(s): random (25) + parsimony (25)\n",
      "  random seed: 2\n",
      "  tip-inner: OFF\n",
      "  pattern compression: ON\n",
      "  per-rate scalers: OFF\n",
      "  site repeats: ON\n",
      "  fast spr radius: AUTO\n",
      "  spr subtree cutoff: 1.000000\n",
      "  branch lengths: proportional (ML estimate, algorithm: NR-FAST)\n",
      "  SIMD kernels: AVX2\n",
      "  parallelization: coarse-grained (auto), PTHREADS (3 threads), thread pinning: OFF\n",
      "\n",
      "[00:00:00] Reading alignment from file: ../results/phyl_trees/all_projects_combined/vamp_mafft.fasta\n",
      "[00:00:00] Loaded alignment with 183 taxa and 2038 sites\n",
      "\n",
      "WARNING: Fully undetermined columns found: 7\n",
      "\n",
      "WARNING: Sequences Thalassomyxa_strain_KibAr_KC779519 and Thalassomyxa_strain_V1ld4_KC779520 are exactly identical!\n",
      "WARNING: Sequences 93ec799fa70cc9a684fdbec04fb423d8_Sim17 and 93ec799fa70cc9a684fdbec04fb423d8_Sim22 are exactly identical!\n",
      "WARNING: Sequences 91b1f98492eba26f9f791cf7949afb56_NH4 and 91b1f98492eba26f9f791cf7949afb56_Sim17 are exactly identical!\n",
      "WARNING: Sequences 91b1f98492eba26f9f791cf7949afb56_NH4 and 91b1f98492eba26f9f791cf7949afb56_Sim22 are exactly identical!\n",
      "WARNING: Sequences 91b1f98492eba26f9f791cf7949afb56_NH4 and 91b1f98492eba26f9f791cf7949afb56_Th38 are exactly identical!\n",
      "WARNING: Sequences 91b1f98492eba26f9f791cf7949afb56_NH4 and 91b1f98492eba26f9f791cf7949afb56_Th40 are exactly identical!\n",
      "WARNING: Sequences fa04010ef186193e6af2567e8e77c9fd_Sim17 and fa04010ef186193e6af2567e8e77c9fd_Sim22 are exactly identical!\n",
      "WARNING: Sequences Placopus_X_GU385615 and Placopus_XX_GU385647 are exactly identical!\n",
      "WARNING: Duplicate sequences found: 8\n",
      "\n",
      "NOTE: Reduced alignment (with duplicates and gap-only sites/taxa removed) \n",
      "NOTE: was saved to: /home/lubo/code/wRajter/vampyrella_2023/notebooks/T2_MLtree.raxml.reduced.phy\n",
      "\n",
      "Alignment comprises 1 partitions and 1485 patterns\n",
      "\n",
      "Partition 0: noname\n",
      "Model: GTR+FO+G4m\n",
      "Alignment sites / patterns: 2031 / 1485\n",
      "Gaps: 32.33 %\n",
      "Invariant sites: 41.36 %\n",
      "\n",
      "\n",
      "NOTE: Binary MSA file created: T2_MLtree.raxml.rba\n",
      "\n",
      "Parallelization scheme autoconfig: 3 worker(s) x 1 thread(s)\n",
      "\n",
      "Parallel reduction/worker buffer size: 1 KB  / 0 KB\n",
      "\n",
      "[00:00:00] Generating 25 random starting tree(s) with 183 taxa\n",
      "[00:00:00] Generating 25 parsimony starting tree(s) with 183 taxa\n",
      "[00:00:00] Data distribution: max. partitions/sites/weight per thread: 1 / 1485 / 23760\n",
      "[00:00:00] Data distribution: max. searches per worker: 17\n",
      "\n",
      "Starting ML tree search with 50 distinct starting trees\n",
      "\n",
      "[00:02:05] [worker #1] ML tree search #2, logLikelihood: -45602.008558\n",
      "[00:02:05] [worker #2] ML tree search #3, logLikelihood: -45607.120366\n",
      "[00:02:14] [worker #0] ML tree search #1, logLikelihood: -45598.523880\n",
      "[00:04:08] [worker #1] ML tree search #5, logLikelihood: -45597.674720\n",
      "[00:04:27] [worker #2] ML tree search #6, logLikelihood: -45600.904169\n",
      "[00:04:31] [worker #0] ML tree search #4, logLikelihood: -45597.393653\n",
      "[00:06:21] [worker #1] ML tree search #8, logLikelihood: -45597.274465\n",
      "[00:06:29] [worker #0] ML tree search #7, logLikelihood: -45598.685906\n",
      "[00:06:33] [worker #2] ML tree search #9, logLikelihood: -45597.891970\n",
      "[00:08:42] [worker #2] ML tree search #12, logLikelihood: -45601.064268\n",
      "[00:08:50] [worker #1] ML tree search #11, logLikelihood: -45597.091811\n",
      "[00:08:52] [worker #0] ML tree search #10, logLikelihood: -45596.343909\n",
      "[00:10:45] [worker #2] ML tree search #15, logLikelihood: -45599.572047\n",
      "[00:11:07] [worker #0] ML tree search #13, logLikelihood: -45604.412971\n",
      "[00:11:09] [worker #1] ML tree search #14, logLikelihood: -45600.792540\n",
      "[00:13:19] [worker #0] ML tree search #16, logLikelihood: -45600.599463\n",
      "[00:13:29] [worker #1] ML tree search #17, logLikelihood: -45597.361195\n",
      "[00:13:30] [worker #2] ML tree search #18, logLikelihood: -45599.134496\n",
      "[00:15:23] [worker #0] ML tree search #19, logLikelihood: -45603.982796\n",
      "[00:15:41] [worker #2] ML tree search #21, logLikelihood: -45598.371858\n",
      "[00:15:45] [worker #1] ML tree search #20, logLikelihood: -45597.138792\n",
      "[00:17:34] [worker #0] ML tree search #22, logLikelihood: -45596.335449\n",
      "[00:17:58] [worker #2] ML tree search #24, logLikelihood: -45597.727585\n",
      "[00:18:04] [worker #1] ML tree search #23, logLikelihood: -45606.929572\n",
      "[00:19:50] [worker #2] ML tree search #27, logLikelihood: -45596.594805\n",
      "[00:19:51] [worker #1] ML tree search #26, logLikelihood: -45597.752833\n",
      "[00:19:54] [worker #0] ML tree search #25, logLikelihood: -45598.899516\n",
      "[00:21:15] [worker #2] ML tree search #30, logLikelihood: -45597.681192\n",
      "[00:21:34] [worker #1] ML tree search #29, logLikelihood: -45600.671089\n",
      "[00:21:38] [worker #0] ML tree search #28, logLikelihood: -45597.764798\n",
      "[00:22:57] [worker #2] ML tree search #33, logLikelihood: -45597.426304\n",
      "[00:23:18] [worker #1] ML tree search #32, logLikelihood: -45603.039861\n",
      "[00:23:28] [worker #0] ML tree search #31, logLikelihood: -45601.447998\n",
      "[00:24:45] [worker #2] ML tree search #36, logLikelihood: -45597.687269\n",
      "[00:25:03] [worker #1] ML tree search #35, logLikelihood: -45597.986655\n",
      "[00:25:20] [worker #0] ML tree search #34, logLikelihood: -45596.646024\n",
      "[00:26:30] [worker #2] ML tree search #39, logLikelihood: -45599.377414\n",
      "[00:27:01] [worker #0] ML tree search #37, logLikelihood: -45598.770785\n",
      "[00:27:02] [worker #1] ML tree search #38, logLikelihood: -45597.594252\n",
      "[00:28:27] [worker #0] ML tree search #40, logLikelihood: -45599.631098\n",
      "[00:28:30] [worker #2] ML tree search #42, logLikelihood: -45596.844256\n",
      "[00:29:00] [worker #1] ML tree search #41, logLikelihood: -45603.161321\n",
      "[00:29:56] [worker #0] ML tree search #43, logLikelihood: -45597.962794\n",
      "[00:30:23] [worker #2] ML tree search #45, logLikelihood: -45597.697977\n",
      "[00:30:34] [worker #1] ML tree search #44, logLikelihood: -45596.189598\n",
      "[00:31:32] [worker #0] ML tree search #46, logLikelihood: -45596.793565\n",
      "[00:31:58] [worker #2] ML tree search #48, logLikelihood: -45598.243831\n",
      "[00:32:18] [worker #1] ML tree search #47, logLikelihood: -45596.599657\n",
      "[00:33:14] [worker #0] ML tree search #49, logLikelihood: -45600.541127\n",
      "[00:33:49] [worker #1] ML tree search #50, logLikelihood: -45599.731702\n",
      "\n",
      "Optimized model parameters:\n",
      "\n",
      "   Partition 0: noname\n",
      "   Rate heterogeneity: GAMMA (4 cats, mean),  alpha: 0.365375 (ML),  weights&rates: (0.250000,0.012052) (0.250000,0.155459) (0.250000,0.690919) (0.250000,3.141570) \n",
      "   Base frequencies (ML): 0.263172 0.189752 0.232549 0.314527 \n",
      "   Substitution rates (ML): 1.174110 3.455438 1.540679 0.891537 4.835594 1.000000 \n",
      "\n",
      "\n",
      "Final LogLikelihood: -45596.189598\n",
      "\n",
      "AIC score: 91936.379195 / AICc score: 92103.756759 / BIC score: 94025.636680\n",
      "Free parameters (model + branch lengths): 372\n",
      "\n",
      "WARNING: Best ML tree contains 13 near-zero branches!\n",
      "\n",
      "Best ML tree with collapsed near-zero branches saved to: /home/lubo/code/wRajter/vampyrella_2023/notebooks/T2_MLtree.raxml.bestTreeCollapsed\n",
      "Best ML tree saved to: /home/lubo/code/wRajter/vampyrella_2023/notebooks/T2_MLtree.raxml.bestTree\n",
      "All ML trees saved to: /home/lubo/code/wRajter/vampyrella_2023/notebooks/T2_MLtree.raxml.mlTrees\n",
      "Optimized model saved to: /home/lubo/code/wRajter/vampyrella_2023/notebooks/T2_MLtree.raxml.bestModel\n",
      "\n",
      "Execution log saved to: /home/lubo/code/wRajter/vampyrella_2023/notebooks/T2_MLtree.raxml.log\n",
      "\n",
      "Analysis started: 11-Dec-2023 11:53:33 / finished: 11-Dec-2023 12:27:23\n",
      "\n",
      "Elapsed time: 2029.973 seconds\n",
      "\n",
      "\n",
      "Output files were moved to: ../results/phyl_trees/all_projects_combined/T2_MLtree\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # Run raxml-ng to compute the maximum likelihood tree and save the output in the specified directory\n",
    "# prefix = 'T2'\n",
    "# output_dir = os.path.join(results_path, 'phyl_trees', project, marker, denoise_method, sim, prefix)\n",
    "\n",
    "\n",
    "prefix = 'T2_MLtree'\n",
    "# output_dir = os.path.join(results_path, 'phyl_trees', 'refrence_alignment_tree', prefix)\n",
    "output_dir = os.path.join(results_path, 'phyl_trees', 'all_projects_combined', prefix)\n",
    "\n",
    "# Check if the output directory exists, create it if not\n",
    "if not os.path.exists(output_dir):\n",
    "    os.makedirs(output_dir)\n",
    "\n",
    "run_raxmlng_tree(alignment = alignment, \n",
    "                 output_dir = output_dir, \n",
    "                 model = model, \n",
    "                 prefix= prefix)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd340fa5-5d22-42cc-a379-97edacec4b3b",
   "metadata": {},
   "source": [
    "## Inferring bootstrap trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9ea4e9ab-833f-45cb-ac6f-6b9fdaa14aee",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "RAxML-NG v. 1.1.0 released on 29.11.2021 by The Exelixis Lab.\n",
      "Developed by: Alexey M. Kozlov and Alexandros Stamatakis.\n",
      "Contributors: Diego Darriba, Tomas Flouri, Benoit Morel, Sarah Lutteropp, Ben Bettisworth.\n",
      "Latest version: https://github.com/amkozlov/raxml-ng\n",
      "Questions/problems/suggestions? Please visit: https://groups.google.com/forum/#!forum/raxml\n",
      "\n",
      "System: 11th Gen Intel(R) Core(TM) i7-11850H @ 2.50GHz, 8 cores, 11 GB RAM\n",
      "\n",
      "RAxML-NG was called at 11-Dec-2023 12:33:00 as follows:\n",
      "\n",
      "raxml-ng --bootstrap --msa ../results/phyl_trees/all_projects_combined/vamp_mafft.fasta --model GTR+G --prefix T3_bootstraps --threads 3 --seed 2\n",
      "\n",
      "Analysis options:\n",
      "  run mode: Bootstrapping\n",
      "  start tree(s): \n",
      "  bootstrap replicates: max: 1000 + bootstopping (autoMRE, cutoff: 0.030000)\n",
      "  random seed: 2\n",
      "  tip-inner: OFF\n",
      "  pattern compression: ON\n",
      "  per-rate scalers: OFF\n",
      "  site repeats: ON\n",
      "  branch lengths: proportional (ML estimate, algorithm: NR-FAST)\n",
      "  SIMD kernels: AVX2\n",
      "  parallelization: coarse-grained (auto), PTHREADS (3 threads), thread pinning: OFF\n",
      "\n",
      "[00:00:00] Reading alignment from file: ../results/phyl_trees/all_projects_combined/vamp_mafft.fasta\n",
      "[00:00:00] Loaded alignment with 183 taxa and 2038 sites\n",
      "\n",
      "WARNING: Fully undetermined columns found: 7\n",
      "\n",
      "WARNING: Sequences Thalassomyxa_strain_KibAr_KC779519 and Thalassomyxa_strain_V1ld4_KC779520 are exactly identical!\n",
      "WARNING: Sequences 93ec799fa70cc9a684fdbec04fb423d8_Sim17 and 93ec799fa70cc9a684fdbec04fb423d8_Sim22 are exactly identical!\n",
      "WARNING: Sequences 91b1f98492eba26f9f791cf7949afb56_NH4 and 91b1f98492eba26f9f791cf7949afb56_Sim17 are exactly identical!\n",
      "WARNING: Sequences 91b1f98492eba26f9f791cf7949afb56_NH4 and 91b1f98492eba26f9f791cf7949afb56_Sim22 are exactly identical!\n",
      "WARNING: Sequences 91b1f98492eba26f9f791cf7949afb56_NH4 and 91b1f98492eba26f9f791cf7949afb56_Th38 are exactly identical!\n",
      "WARNING: Sequences 91b1f98492eba26f9f791cf7949afb56_NH4 and 91b1f98492eba26f9f791cf7949afb56_Th40 are exactly identical!\n",
      "WARNING: Sequences fa04010ef186193e6af2567e8e77c9fd_Sim17 and fa04010ef186193e6af2567e8e77c9fd_Sim22 are exactly identical!\n",
      "WARNING: Sequences Placopus_X_GU385615 and Placopus_XX_GU385647 are exactly identical!\n",
      "WARNING: Duplicate sequences found: 8\n",
      "\n",
      "NOTE: Reduced alignment (with duplicates and gap-only sites/taxa removed) \n",
      "NOTE: was saved to: /home/lubo/code/wRajter/vampyrella_2023/notebooks/T3_bootstraps.raxml.reduced.phy\n",
      "\n",
      "Alignment comprises 1 partitions and 1485 patterns\n",
      "\n",
      "Partition 0: noname\n",
      "Model: GTR+FO+G4m\n",
      "Alignment sites / patterns: 2031 / 1485\n",
      "Gaps: 32.33 %\n",
      "Invariant sites: 41.36 %\n",
      "\n",
      "\n",
      "NOTE: Binary MSA file created: T3_bootstraps.raxml.rba\n",
      "\n",
      "Parallelization scheme autoconfig: 3 worker(s) x 1 thread(s)\n",
      "\n",
      "Parallel reduction/worker buffer size: 1 KB  / 0 KB\n",
      "\n",
      "[00:00:00] Data distribution: max. partitions/sites/weight per thread: 1 / 1485 / 23760\n",
      "[00:00:00] Data distribution: max. searches per worker: 334\n",
      "[00:00:00] Starting bootstrapping analysis with 1000 replicates.\n",
      "\n",
      "[00:01:21] [worker #1] Bootstrap tree #2, logLikelihood: -44869.031379\n",
      "[00:01:57] [worker #0] Bootstrap tree #1, logLikelihood: -44429.424950\n",
      "[00:02:03] [worker #2] Bootstrap tree #3, logLikelihood: -41538.998050\n",
      "[00:02:46] [worker #1] Bootstrap tree #5, logLikelihood: -44154.887935\n",
      "[00:03:25] [worker #0] Bootstrap tree #4, logLikelihood: -46735.432134\n",
      "[00:03:36] [worker #2] Bootstrap tree #6, logLikelihood: -44204.626810\n",
      "[00:04:18] [worker #1] Bootstrap tree #8, logLikelihood: -45237.013310\n",
      "[00:04:53] [worker #0] Bootstrap tree #7, logLikelihood: -45025.108039\n",
      "[00:05:04] [worker #2] Bootstrap tree #9, logLikelihood: -44321.401318\n",
      "[00:05:55] [worker #1] Bootstrap tree #11, logLikelihood: -44991.322776\n",
      "[00:06:28] [worker #0] Bootstrap tree #10, logLikelihood: -46297.575597\n",
      "[00:06:42] [worker #2] Bootstrap tree #12, logLikelihood: -45301.455693\n",
      "[00:07:24] [worker #1] Bootstrap tree #14, logLikelihood: -44979.176957\n",
      "[00:08:09] [worker #0] Bootstrap tree #13, logLikelihood: -45527.835447\n",
      "[00:08:44] [worker #1] Bootstrap tree #17, logLikelihood: -44629.531396\n",
      "[00:08:47] [worker #2] Bootstrap tree #15, logLikelihood: -44525.196343\n",
      "[00:09:33] [worker #0] Bootstrap tree #16, logLikelihood: -44171.314032\n",
      "[00:10:03] [worker #1] Bootstrap tree #20, logLikelihood: -45106.576272\n",
      "[00:10:16] [worker #2] Bootstrap tree #18, logLikelihood: -45043.556774\n",
      "[00:10:53] [worker #0] Bootstrap tree #19, logLikelihood: -46861.191130\n",
      "[00:11:30] [worker #1] Bootstrap tree #23, logLikelihood: -45612.785555\n",
      "[00:11:43] [worker #2] Bootstrap tree #21, logLikelihood: -44679.631491\n",
      "[00:12:23] [worker #0] Bootstrap tree #22, logLikelihood: -43531.625920\n",
      "[00:12:53] [worker #1] Bootstrap tree #26, logLikelihood: -45836.364393\n",
      "[00:13:18] [worker #2] Bootstrap tree #24, logLikelihood: -45679.234231\n",
      "[00:14:28] [worker #1] Bootstrap tree #29, logLikelihood: -46210.583345\n",
      "[00:14:40] [worker #0] Bootstrap tree #25, logLikelihood: -45534.833653\n",
      "[00:14:50] [worker #2] Bootstrap tree #27, logLikelihood: -44176.573705\n",
      "[00:15:53] [worker #1] Bootstrap tree #32, logLikelihood: -44201.235241\n",
      "[00:16:11] [worker #0] Bootstrap tree #28, logLikelihood: -45658.176384\n",
      "[00:16:14] [worker #2] Bootstrap tree #30, logLikelihood: -43904.940985\n",
      "[00:17:23] [worker #1] Bootstrap tree #35, logLikelihood: -44544.794288\n",
      "[00:18:01] [worker #0] Bootstrap tree #31, logLikelihood: -45748.649649\n",
      "[00:18:15] [worker #2] Bootstrap tree #33, logLikelihood: -44704.875511\n",
      "[00:19:14] [worker #1] Bootstrap tree #38, logLikelihood: -46604.386623\n",
      "[00:19:41] [worker #2] Bootstrap tree #36, logLikelihood: -44537.568006\n",
      "[00:20:49] [worker #0] Bootstrap tree #34, logLikelihood: -44246.756522\n",
      "[00:21:05] [worker #1] Bootstrap tree #41, logLikelihood: -46216.407663\n",
      "[00:21:16] [worker #2] Bootstrap tree #39, logLikelihood: -44930.794227\n",
      "[00:22:15] [worker #0] Bootstrap tree #37, logLikelihood: -47306.143572\n",
      "[00:23:36] [worker #2] Bootstrap tree #42, logLikelihood: -45014.811279\n",
      "[00:24:08] [worker #1] Bootstrap tree #44, logLikelihood: -47618.101763\n",
      "[00:24:48] [worker #0] Bootstrap tree #40, logLikelihood: -46813.909839\n",
      "[00:25:24] [worker #2] Bootstrap tree #45, logLikelihood: -47038.344024\n",
      "[00:25:38] [worker #1] Bootstrap tree #47, logLikelihood: -45887.572007\n",
      "[00:26:18] [worker #0] Bootstrap tree #43, logLikelihood: -47532.942698\n",
      "[00:26:59] [worker #2] Bootstrap tree #48, logLikelihood: -46008.736846\n",
      "[00:27:03] [worker #1] Bootstrap tree #50, logLikelihood: -47125.038846\n",
      "[00:28:06] [worker #0] Bootstrap tree #46, logLikelihood: -45595.041661\n",
      "[00:29:28] [worker #0] Bootstrap tree #49, logLikelihood: -45921.936539\n",
      "[00:32:04] [worker #1] Bootstrap tree #53, logLikelihood: -45731.780371\n",
      "[00:32:06] [worker #0] Bootstrap tree #52, logLikelihood: -44387.628081\n",
      "[00:32:36] [worker #2] Bootstrap tree #51, logLikelihood: -47714.321316\n",
      "[00:34:05] [worker #1] Bootstrap tree #56, logLikelihood: -44424.284085\n",
      "[00:34:26] [worker #0] Bootstrap tree #55, logLikelihood: -45448.171458\n",
      "[00:35:29] [worker #2] Bootstrap tree #54, logLikelihood: -48083.665035\n",
      "[00:35:32] [worker #1] Bootstrap tree #59, logLikelihood: -44559.826016\n",
      "[00:36:16] [worker #0] Bootstrap tree #58, logLikelihood: -47896.193665\n",
      "[00:36:57] [worker #1] Bootstrap tree #62, logLikelihood: -43725.376548\n",
      "[00:37:23] [worker #2] Bootstrap tree #57, logLikelihood: -45611.242572\n",
      "[00:38:09] [worker #0] Bootstrap tree #61, logLikelihood: -45808.721200\n",
      "[00:38:31] [worker #1] Bootstrap tree #65, logLikelihood: -46167.646496\n",
      "[00:38:59] [worker #2] Bootstrap tree #60, logLikelihood: -43314.818462\n",
      "[00:39:39] [worker #0] Bootstrap tree #64, logLikelihood: -43609.478242\n",
      "[00:40:21] [worker #2] Bootstrap tree #63, logLikelihood: -46133.117314\n",
      "[00:40:27] [worker #1] Bootstrap tree #68, logLikelihood: -46675.947752\n",
      "[00:41:06] [worker #0] Bootstrap tree #67, logLikelihood: -46240.577436\n",
      "[00:41:45] [worker #2] Bootstrap tree #66, logLikelihood: -46440.108293\n",
      "[00:42:32] [worker #0] Bootstrap tree #70, logLikelihood: -43997.420310\n",
      "[00:42:40] [worker #1] Bootstrap tree #71, logLikelihood: -43330.590882\n",
      "[00:43:09] [worker #2] Bootstrap tree #69, logLikelihood: -45718.229156\n",
      "[00:44:08] [worker #1] Bootstrap tree #74, logLikelihood: -44211.336475\n",
      "[00:44:49] [worker #0] Bootstrap tree #73, logLikelihood: -44469.484198\n",
      "[00:44:53] [worker #2] Bootstrap tree #72, logLikelihood: -47126.196673\n",
      "[00:45:48] [worker #1] Bootstrap tree #77, logLikelihood: -45872.251446\n",
      "[00:46:33] [worker #2] Bootstrap tree #75, logLikelihood: -43704.957265\n",
      "[00:46:53] [worker #0] Bootstrap tree #76, logLikelihood: -44422.746710\n",
      "[00:47:18] [worker #1] Bootstrap tree #80, logLikelihood: -46388.293324\n",
      "[00:48:03] [worker #2] Bootstrap tree #78, logLikelihood: -44628.717472\n",
      "[00:48:25] [worker #0] Bootstrap tree #79, logLikelihood: -45426.545674\n",
      "[00:49:01] [worker #1] Bootstrap tree #83, logLikelihood: -45407.401334\n",
      "[00:49:38] [worker #2] Bootstrap tree #81, logLikelihood: -44447.944043\n",
      "[00:50:17] [worker #0] Bootstrap tree #82, logLikelihood: -43902.060304\n",
      "[00:51:01] [worker #1] Bootstrap tree #86, logLikelihood: -43619.057366\n",
      "[00:51:10] [worker #2] Bootstrap tree #84, logLikelihood: -45768.258806\n",
      "[00:51:39] [worker #0] Bootstrap tree #85, logLikelihood: -45797.572890\n",
      "[00:52:23] [worker #1] Bootstrap tree #89, logLikelihood: -45121.799412\n",
      "[00:52:38] [worker #2] Bootstrap tree #87, logLikelihood: -47035.630365\n",
      "[00:53:09] [worker #0] Bootstrap tree #88, logLikelihood: -46073.533352\n",
      "[00:54:10] [worker #1] Bootstrap tree #92, logLikelihood: -45195.068381\n",
      "[00:54:22] [worker #2] Bootstrap tree #90, logLikelihood: -46854.852618\n",
      "[00:54:47] [worker #0] Bootstrap tree #91, logLikelihood: -43763.088203\n",
      "[00:55:35] [worker #1] Bootstrap tree #95, logLikelihood: -44719.735699\n",
      "[00:56:07] [worker #0] Bootstrap tree #94, logLikelihood: -45656.065983\n",
      "[00:56:31] [worker #2] Bootstrap tree #93, logLikelihood: -45955.909157\n",
      "[00:57:06] [worker #1] Bootstrap tree #98, logLikelihood: -46093.966361\n",
      "[00:57:26] [worker #0] Bootstrap tree #97, logLikelihood: -46394.855792\n",
      "[00:57:48] [worker #2] Bootstrap tree #96, logLikelihood: -44217.612480\n",
      "[00:59:15] [worker #2] Bootstrap tree #99, logLikelihood: -46237.450657\n",
      "[00:59:20] [worker #0] Bootstrap tree #100, logLikelihood: -45769.336790\n",
      "[01:00:53] [worker #2] Bootstrap tree #102, logLikelihood: -46643.794263\n",
      "[01:00:57] [worker #1] Bootstrap tree #101, logLikelihood: -45060.905577\n",
      "[01:01:36] [worker #0] Bootstrap tree #103, logLikelihood: -44779.558101\n",
      "[01:02:32] [worker #1] Bootstrap tree #104, logLikelihood: -44171.522710\n",
      "[01:02:50] [worker #2] Bootstrap tree #105, logLikelihood: -48325.148692\n",
      "[01:03:17] [worker #0] Bootstrap tree #106, logLikelihood: -46950.406406\n",
      "[01:04:38] [worker #1] Bootstrap tree #107, logLikelihood: -47142.940400\n",
      "[01:04:38] [worker #2] Bootstrap tree #108, logLikelihood: -45718.372267\n",
      "[01:05:10] [worker #0] Bootstrap tree #109, logLikelihood: -46929.786037\n",
      "[01:06:17] [worker #2] Bootstrap tree #111, logLikelihood: -45455.684451\n",
      "[01:06:44] [worker #1] Bootstrap tree #110, logLikelihood: -46638.093168\n",
      "[01:07:17] [worker #0] Bootstrap tree #112, logLikelihood: -44502.779576\n",
      "[01:08:04] [worker #1] Bootstrap tree #113, logLikelihood: -43909.251584\n",
      "[01:08:27] [worker #2] Bootstrap tree #114, logLikelihood: -45947.884922\n",
      "[01:08:46] [worker #0] Bootstrap tree #115, logLikelihood: -45615.587938\n",
      "[01:09:42] [worker #1] Bootstrap tree #116, logLikelihood: -46674.319529\n",
      "[01:09:58] [worker #2] Bootstrap tree #117, logLikelihood: -46743.126066\n",
      "[01:10:10] [worker #0] Bootstrap tree #118, logLikelihood: -44549.121374\n",
      "[01:11:49] [worker #0] Bootstrap tree #121, logLikelihood: -45269.944828\n",
      "[01:11:49] [worker #1] Bootstrap tree #119, logLikelihood: -45217.840791\n",
      "[01:12:42] [worker #2] Bootstrap tree #120, logLikelihood: -44409.551787\n",
      "[01:13:09] [worker #0] Bootstrap tree #124, logLikelihood: -44634.101289\n",
      "[01:13:43] [worker #1] Bootstrap tree #122, logLikelihood: -45310.456691\n",
      "[01:14:17] [worker #2] Bootstrap tree #123, logLikelihood: -44151.347272\n",
      "[01:14:30] [worker #0] Bootstrap tree #127, logLikelihood: -45112.575211\n",
      "[01:15:31] [worker #1] Bootstrap tree #125, logLikelihood: -45572.931385\n",
      "[01:16:16] [worker #2] Bootstrap tree #126, logLikelihood: -46210.940765\n",
      "[01:16:25] [worker #0] Bootstrap tree #130, logLikelihood: -45455.489501\n",
      "[01:16:54] [worker #1] Bootstrap tree #128, logLikelihood: -45364.215847\n",
      "[01:17:53] [worker #2] Bootstrap tree #129, logLikelihood: -44506.311842\n",
      "[01:17:57] [worker #0] Bootstrap tree #133, logLikelihood: -45948.795965\n",
      "[01:18:13] [worker #1] Bootstrap tree #131, logLikelihood: -45866.871029\n",
      "[01:19:23] [worker #2] Bootstrap tree #132, logLikelihood: -46964.073228\n",
      "[01:19:40] [worker #1] Bootstrap tree #134, logLikelihood: -45322.340182\n",
      "[01:20:14] [worker #0] Bootstrap tree #136, logLikelihood: -47276.021862\n",
      "[01:21:07] [worker #1] Bootstrap tree #137, logLikelihood: -45303.830249\n",
      "[01:21:18] [worker #2] Bootstrap tree #135, logLikelihood: -45563.504664\n",
      "[01:22:25] [worker #0] Bootstrap tree #139, logLikelihood: -44784.750878\n",
      "[01:22:45] [worker #1] Bootstrap tree #140, logLikelihood: -46190.482846\n",
      "[01:22:51] [worker #2] Bootstrap tree #138, logLikelihood: -44346.351698\n",
      "[01:24:12] [worker #1] Bootstrap tree #143, logLikelihood: -44738.714649\n",
      "[01:24:14] [worker #2] Bootstrap tree #141, logLikelihood: -46310.800240\n",
      "[01:25:13] [worker #0] Bootstrap tree #142, logLikelihood: -46331.574920\n",
      "[01:25:34] [worker #2] Bootstrap tree #144, logLikelihood: -45771.263910\n",
      "[01:25:36] [worker #1] Bootstrap tree #146, logLikelihood: -45442.280726\n",
      "[01:26:35] [worker #0] Bootstrap tree #145, logLikelihood: -46367.776526\n",
      "[01:27:04] [worker #1] Bootstrap tree #149, logLikelihood: -43638.573736\n",
      "[01:27:05] [worker #2] Bootstrap tree #147, logLikelihood: -43639.021611\n",
      "[01:28:03] [worker #0] Bootstrap tree #148, logLikelihood: -44905.149372\n",
      "[01:28:30] [worker #2] Bootstrap tree #150, logLikelihood: -44927.534143\n",
      "[01:30:02] [worker #1] Bootstrap tree #152, logLikelihood: -44969.680442\n",
      "[01:30:09] [worker #0] Bootstrap tree #151, logLikelihood: -45641.720618\n",
      "[01:30:16] [worker #2] Bootstrap tree #153, logLikelihood: -44427.894274\n",
      "[01:31:30] [worker #1] Bootstrap tree #155, logLikelihood: -43395.743538\n",
      "[01:31:49] [worker #0] Bootstrap tree #154, logLikelihood: -46092.499985\n",
      "[01:32:50] [worker #2] Bootstrap tree #156, logLikelihood: -45493.841692\n",
      "[01:33:06] [worker #1] Bootstrap tree #158, logLikelihood: -45966.873038\n",
      "[01:33:56] [worker #0] Bootstrap tree #157, logLikelihood: -44873.113219\n",
      "[01:34:25] [worker #2] Bootstrap tree #159, logLikelihood: -45903.024097\n",
      "[01:34:34] [worker #1] Bootstrap tree #161, logLikelihood: -46177.078709\n",
      "[01:35:29] [worker #0] Bootstrap tree #160, logLikelihood: -45380.734686\n",
      "[01:35:44] [worker #2] Bootstrap tree #162, logLikelihood: -47179.517868\n",
      "[01:36:02] [worker #1] Bootstrap tree #164, logLikelihood: -45460.767460\n",
      "[01:37:09] [worker #0] Bootstrap tree #163, logLikelihood: -45655.299638\n",
      "[01:37:35] [worker #1] Bootstrap tree #167, logLikelihood: -46002.659684\n",
      "[01:37:57] [worker #2] Bootstrap tree #165, logLikelihood: -44721.787267\n",
      "[01:38:46] [worker #0] Bootstrap tree #166, logLikelihood: -48344.064717\n",
      "[01:38:58] [worker #1] Bootstrap tree #170, logLikelihood: -47946.798884\n",
      "[01:39:27] [worker #2] Bootstrap tree #168, logLikelihood: -43064.795699\n",
      "[01:40:14] [worker #0] Bootstrap tree #169, logLikelihood: -44493.175659\n",
      "[01:40:59] [worker #1] Bootstrap tree #173, logLikelihood: -45155.363176\n",
      "[01:41:00] [worker #2] Bootstrap tree #171, logLikelihood: -43353.391701\n",
      "[01:41:32] [worker #0] Bootstrap tree #172, logLikelihood: -43460.830283\n",
      "[01:42:24] [worker #1] Bootstrap tree #176, logLikelihood: -43674.599051\n",
      "[01:42:43] [worker #2] Bootstrap tree #174, logLikelihood: -45583.341217\n",
      "[01:42:58] [worker #0] Bootstrap tree #175, logLikelihood: -46161.036280\n",
      "[01:43:48] [worker #1] Bootstrap tree #179, logLikelihood: -46365.535876\n",
      "[01:44:30] [worker #0] Bootstrap tree #178, logLikelihood: -42001.628994\n",
      "[01:44:39] [worker #2] Bootstrap tree #177, logLikelihood: -47294.258698\n",
      "[01:45:29] [worker #1] Bootstrap tree #182, logLikelihood: -43834.513718\n",
      "[01:45:59] [worker #0] Bootstrap tree #181, logLikelihood: -43931.272006\n",
      "[01:46:08] [worker #2] Bootstrap tree #180, logLikelihood: -44109.529064\n",
      "[01:46:52] [worker #1] Bootstrap tree #185, logLikelihood: -44982.429958\n",
      "[01:47:25] [worker #0] Bootstrap tree #184, logLikelihood: -46705.403935\n",
      "[01:47:39] [worker #2] Bootstrap tree #183, logLikelihood: -44802.443033\n",
      "[01:48:25] [worker #1] Bootstrap tree #188, logLikelihood: -43506.686309\n",
      "[01:48:50] [worker #0] Bootstrap tree #187, logLikelihood: -46529.736956\n",
      "[01:49:26] [worker #2] Bootstrap tree #186, logLikelihood: -43856.064104\n",
      "[01:50:01] [worker #1] Bootstrap tree #191, logLikelihood: -47119.936152\n",
      "[01:50:59] [worker #2] Bootstrap tree #189, logLikelihood: -48721.986229\n",
      "[01:51:01] [worker #0] Bootstrap tree #190, logLikelihood: -44424.116488\n",
      "[01:51:32] [worker #1] Bootstrap tree #194, logLikelihood: -45691.516933\n",
      "[01:52:37] [worker #0] Bootstrap tree #193, logLikelihood: -45928.531236\n",
      "[01:52:42] [worker #2] Bootstrap tree #192, logLikelihood: -45767.754347\n",
      "[01:53:05] [worker #1] Bootstrap tree #197, logLikelihood: -43403.636958\n",
      "[01:54:13] [worker #2] Bootstrap tree #195, logLikelihood: -44672.209016\n",
      "[01:54:21] [worker #0] Bootstrap tree #196, logLikelihood: -46292.130843\n",
      "[01:54:50] [worker #1] Bootstrap tree #200, logLikelihood: -46974.123493\n",
      "[01:55:48] [worker #2] Bootstrap tree #198, logLikelihood: -44616.501406\n",
      "[01:56:25] [worker #0] Bootstrap tree #199, logLikelihood: -46121.924649\n",
      "[01:58:02] [worker #1] Bootstrap tree #203, logLikelihood: -46790.867034\n",
      "[01:58:03] [worker #2] Bootstrap tree #201, logLikelihood: -42727.996642\n",
      "[01:58:12] [worker #0] Bootstrap tree #202, logLikelihood: -44355.078580\n",
      "[01:59:50] [worker #1] Bootstrap tree #206, logLikelihood: -45855.716293\n",
      "[01:59:54] [worker #0] Bootstrap tree #205, logLikelihood: -47205.755777\n",
      "[02:00:04] [worker #2] Bootstrap tree #204, logLikelihood: -44865.800371\n",
      "[02:01:46] [worker #0] Bootstrap tree #208, logLikelihood: -48472.340489\n",
      "[02:01:54] [worker #2] Bootstrap tree #207, logLikelihood: -46888.384573\n",
      "[02:02:01] [worker #1] Bootstrap tree #209, logLikelihood: -44518.037895\n",
      "[02:03:29] [worker #2] Bootstrap tree #210, logLikelihood: -46273.805472\n",
      "[02:03:37] [worker #0] Bootstrap tree #211, logLikelihood: -42572.633437\n",
      "[02:03:40] [worker #1] Bootstrap tree #212, logLikelihood: -45686.110434\n",
      "[02:05:10] [worker #1] Bootstrap tree #215, logLikelihood: -44583.335070\n",
      "[02:05:17] [worker #0] Bootstrap tree #214, logLikelihood: -46018.791825\n",
      "[02:05:49] [worker #2] Bootstrap tree #213, logLikelihood: -42574.449611\n",
      "[02:06:52] [worker #1] Bootstrap tree #218, logLikelihood: -44260.317940\n",
      "[02:06:54] [worker #0] Bootstrap tree #217, logLikelihood: -44520.314884\n",
      "[02:07:07] [worker #2] Bootstrap tree #216, logLikelihood: -44507.690091\n",
      "[02:08:16] [worker #1] Bootstrap tree #221, logLikelihood: -45583.589401\n",
      "[02:08:29] [worker #0] Bootstrap tree #220, logLikelihood: -46670.418458\n",
      "[02:08:35] [worker #2] Bootstrap tree #219, logLikelihood: -46615.477220\n",
      "[02:10:28] [worker #2] Bootstrap tree #222, logLikelihood: -46825.284870\n",
      "[02:10:38] [worker #1] Bootstrap tree #224, logLikelihood: -42769.874880\n",
      "[02:11:08] [worker #0] Bootstrap tree #223, logLikelihood: -42402.304929\n",
      "[02:12:05] [worker #1] Bootstrap tree #227, logLikelihood: -42809.445991\n",
      "[02:12:18] [worker #2] Bootstrap tree #225, logLikelihood: -48064.087962\n",
      "[02:12:47] [worker #0] Bootstrap tree #226, logLikelihood: -45130.365689\n",
      "[02:13:45] [worker #1] Bootstrap tree #230, logLikelihood: -44105.301958\n",
      "[02:14:13] [worker #2] Bootstrap tree #228, logLikelihood: -46321.485900\n",
      "[02:14:22] [worker #0] Bootstrap tree #229, logLikelihood: -45322.290340\n",
      "[02:15:20] [worker #1] Bootstrap tree #233, logLikelihood: -44754.091073\n",
      "[02:15:39] [worker #2] Bootstrap tree #231, logLikelihood: -45920.973107\n",
      "[02:16:40] [worker #0] Bootstrap tree #232, logLikelihood: -43627.457789\n",
      "[02:16:58] [worker #2] Bootstrap tree #234, logLikelihood: -47351.743157\n",
      "[02:17:13] [worker #1] Bootstrap tree #236, logLikelihood: -46983.109789\n",
      "[02:18:00] [worker #0] Bootstrap tree #235, logLikelihood: -45376.700867\n",
      "[02:18:20] [worker #2] Bootstrap tree #237, logLikelihood: -44289.374287\n",
      "[02:18:46] [worker #1] Bootstrap tree #239, logLikelihood: -45724.848607\n",
      "[02:19:19] [worker #0] Bootstrap tree #238, logLikelihood: -46291.172637\n",
      "[02:19:42] [worker #2] Bootstrap tree #240, logLikelihood: -44282.490620\n",
      "[02:20:11] [worker #1] Bootstrap tree #242, logLikelihood: -42443.213934\n",
      "[02:20:49] [worker #0] Bootstrap tree #241, logLikelihood: -45993.665537\n",
      "[02:21:00] [worker #2] Bootstrap tree #243, logLikelihood: -44638.434198\n",
      "[02:21:30] [worker #1] Bootstrap tree #245, logLikelihood: -42948.548372\n",
      "[02:22:19] [worker #0] Bootstrap tree #244, logLikelihood: -45686.464575\n",
      "[02:22:20] [worker #2] Bootstrap tree #246, logLikelihood: -45957.859581\n",
      "[02:23:16] [worker #1] Bootstrap tree #248, logLikelihood: -45751.023936\n",
      "[02:23:45] [worker #2] Bootstrap tree #249, logLikelihood: -46184.021964\n",
      "[02:24:14] [worker #0] Bootstrap tree #247, logLikelihood: -44748.863618\n",
      "[02:25:47] [worker #0] Bootstrap tree #250, logLikelihood: -46974.542696\n",
      "[02:27:17] [worker #2] Bootstrap tree #252, logLikelihood: -46016.312140\n",
      "[02:27:19] [worker #1] Bootstrap tree #251, logLikelihood: -48716.557289\n",
      "[02:27:35] [worker #0] Bootstrap tree #253, logLikelihood: -45442.691758\n",
      "[02:28:45] [worker #1] Bootstrap tree #254, logLikelihood: -43339.082497\n",
      "[02:28:54] [worker #2] Bootstrap tree #255, logLikelihood: -45190.087371\n",
      "[02:28:59] [worker #0] Bootstrap tree #256, logLikelihood: -44494.123381\n",
      "[02:30:52] [worker #1] Bootstrap tree #257, logLikelihood: -43914.440908\n",
      "[02:30:55] [worker #0] Bootstrap tree #259, logLikelihood: -46617.613653\n",
      "[02:31:00] [worker #2] Bootstrap tree #258, logLikelihood: -42637.294232\n",
      "[02:32:27] [worker #1] Bootstrap tree #260, logLikelihood: -44747.219850\n",
      "[02:32:43] [worker #0] Bootstrap tree #262, logLikelihood: -47460.675955\n",
      "[02:33:20] [worker #2] Bootstrap tree #261, logLikelihood: -46024.340548\n",
      "[02:34:07] [worker #0] Bootstrap tree #265, logLikelihood: -45802.820880\n",
      "[02:34:13] [worker #1] Bootstrap tree #263, logLikelihood: -46267.424767\n",
      "[02:35:09] [worker #2] Bootstrap tree #264, logLikelihood: -44852.423608\n",
      "[02:35:31] [worker #0] Bootstrap tree #268, logLikelihood: -46333.944212\n",
      "[02:35:45] [worker #1] Bootstrap tree #266, logLikelihood: -44138.203109\n",
      "[02:37:00] [worker #2] Bootstrap tree #267, logLikelihood: -43213.287046\n",
      "[02:37:36] [worker #0] Bootstrap tree #271, logLikelihood: -48546.761421\n",
      "[02:37:43] [worker #1] Bootstrap tree #269, logLikelihood: -45911.767559\n",
      "[02:38:50] [worker #2] Bootstrap tree #270, logLikelihood: -48536.359252\n",
      "[02:38:58] [worker #1] Bootstrap tree #272, logLikelihood: -43165.372971\n",
      "[02:39:08] [worker #0] Bootstrap tree #274, logLikelihood: -45715.679120\n",
      "[02:40:30] [worker #2] Bootstrap tree #273, logLikelihood: -46174.935748\n",
      "[02:40:37] [worker #1] Bootstrap tree #275, logLikelihood: -44725.821401\n",
      "[02:40:42] [worker #0] Bootstrap tree #277, logLikelihood: -45325.045190\n",
      "[02:42:03] [worker #2] Bootstrap tree #276, logLikelihood: -45627.918806\n",
      "[02:42:08] [worker #1] Bootstrap tree #278, logLikelihood: -46308.599886\n",
      "[02:42:13] [worker #0] Bootstrap tree #280, logLikelihood: -45652.633633\n",
      "[02:43:38] [worker #1] Bootstrap tree #281, logLikelihood: -44612.293389\n",
      "[02:43:53] [worker #0] Bootstrap tree #283, logLikelihood: -43694.012784\n",
      "[02:44:01] [worker #2] Bootstrap tree #279, logLikelihood: -44619.601574\n",
      "[02:45:07] [worker #1] Bootstrap tree #284, logLikelihood: -42819.104580\n",
      "[02:45:18] [worker #0] Bootstrap tree #286, logLikelihood: -47053.485248\n",
      "[02:45:39] [worker #2] Bootstrap tree #282, logLikelihood: -46842.824762\n",
      "[02:46:44] [worker #1] Bootstrap tree #287, logLikelihood: -46562.700167\n",
      "[02:46:57] [worker #0] Bootstrap tree #289, logLikelihood: -44996.153256\n",
      "[02:47:09] [worker #2] Bootstrap tree #285, logLikelihood: -44853.713756\n",
      "[02:48:29] [worker #1] Bootstrap tree #290, logLikelihood: -45023.998478\n",
      "[02:48:40] [worker #0] Bootstrap tree #292, logLikelihood: -46019.184742\n",
      "[02:48:44] [worker #2] Bootstrap tree #288, logLikelihood: -45904.610861\n",
      "[02:50:25] [worker #2] Bootstrap tree #291, logLikelihood: -46466.911463\n",
      "[02:50:31] [worker #1] Bootstrap tree #293, logLikelihood: -45131.274256\n",
      "[02:51:01] [worker #0] Bootstrap tree #295, logLikelihood: -43409.690163\n",
      "[02:51:58] [worker #2] Bootstrap tree #294, logLikelihood: -45461.658137\n",
      "[02:52:15] [worker #1] Bootstrap tree #296, logLikelihood: -45363.717688\n",
      "[02:53:35] [worker #0] Bootstrap tree #298, logLikelihood: -46121.227662\n",
      "[02:53:42] [worker #2] Bootstrap tree #297, logLikelihood: -43899.444928\n",
      "[02:53:53] [worker #1] Bootstrap tree #299, logLikelihood: -43929.156247\n",
      "[02:55:29] [worker #2] Bootstrap tree #300, logLikelihood: -44901.754384\n",
      "[02:57:15] [worker #1] Bootstrap tree #302, logLikelihood: -46593.246857\n",
      "[02:57:25] [worker #2] Bootstrap tree #303, logLikelihood: -44544.890764\n",
      "[02:58:13] [worker #0] Bootstrap tree #301, logLikelihood: -46682.557092\n",
      "[02:59:19] [worker #1] Bootstrap tree #305, logLikelihood: -44118.228442\n",
      "[02:59:31] [worker #2] Bootstrap tree #306, logLikelihood: -44833.930350\n",
      "[03:00:05] [worker #0] Bootstrap tree #304, logLikelihood: -47559.381845\n",
      "[03:01:01] [worker #1] Bootstrap tree #308, logLikelihood: -44734.988972\n",
      "[03:01:30] [worker #2] Bootstrap tree #309, logLikelihood: -43723.751419\n",
      "[03:01:51] [worker #0] Bootstrap tree #307, logLikelihood: -44221.362371\n",
      "[03:02:43] [worker #1] Bootstrap tree #311, logLikelihood: -46055.149306\n",
      "[03:03:12] [worker #2] Bootstrap tree #312, logLikelihood: -43880.757719\n",
      "[03:03:52] [worker #0] Bootstrap tree #310, logLikelihood: -46708.255666\n",
      "[03:04:51] [worker #1] Bootstrap tree #314, logLikelihood: -44928.831647\n",
      "[03:05:17] [worker #2] Bootstrap tree #315, logLikelihood: -43844.187886\n",
      "[03:06:47] [worker #1] Bootstrap tree #317, logLikelihood: -47929.253214\n",
      "[03:07:01] [worker #0] Bootstrap tree #313, logLikelihood: -45355.290565\n",
      "[03:07:05] [worker #2] Bootstrap tree #318, logLikelihood: -43650.379070\n",
      "[03:08:36] [worker #1] Bootstrap tree #320, logLikelihood: -46994.490734\n",
      "[03:08:43] [worker #0] Bootstrap tree #316, logLikelihood: -43867.625551\n",
      "[03:08:57] [worker #2] Bootstrap tree #321, logLikelihood: -46066.069730\n",
      "[03:10:27] [worker #1] Bootstrap tree #323, logLikelihood: -43348.636904\n",
      "[03:10:31] [worker #0] Bootstrap tree #319, logLikelihood: -45849.012991\n",
      "[03:10:54] [worker #2] Bootstrap tree #324, logLikelihood: -47829.586797\n",
      "[03:12:06] [worker #1] Bootstrap tree #326, logLikelihood: -43751.883186\n",
      "[03:12:29] [worker #2] Bootstrap tree #327, logLikelihood: -43126.794411\n",
      "[03:13:06] [worker #0] Bootstrap tree #322, logLikelihood: -44345.482080\n",
      "[03:13:48] [worker #1] Bootstrap tree #329, logLikelihood: -46789.709687\n",
      "[03:14:07] [worker #2] Bootstrap tree #330, logLikelihood: -45141.989574\n",
      "[03:14:53] [worker #0] Bootstrap tree #325, logLikelihood: -44557.051434\n",
      "[03:15:25] [worker #1] Bootstrap tree #332, logLikelihood: -45170.997468\n",
      "[03:15:35] [worker #2] Bootstrap tree #333, logLikelihood: -44837.828027\n",
      "[03:16:47] [worker #0] Bootstrap tree #328, logLikelihood: -45276.636446\n",
      "[03:17:08] [worker #2] Bootstrap tree #336, logLikelihood: -44812.870004\n",
      "[03:17:11] [worker #1] Bootstrap tree #335, logLikelihood: -44861.071140\n",
      "[03:18:16] [worker #0] Bootstrap tree #331, logLikelihood: -45181.686971\n",
      "[03:18:43] [worker #1] Bootstrap tree #338, logLikelihood: -45482.847845\n",
      "[03:18:52] [worker #2] Bootstrap tree #339, logLikelihood: -44923.849455\n",
      "[03:19:56] [worker #0] Bootstrap tree #334, logLikelihood: -45356.025356\n",
      "[03:20:33] [worker #2] Bootstrap tree #342, logLikelihood: -44707.189410\n",
      "[03:20:40] [worker #1] Bootstrap tree #341, logLikelihood: -46249.822773\n",
      "[03:21:22] [worker #0] Bootstrap tree #337, logLikelihood: -45975.642403\n",
      "[03:21:54] [worker #2] Bootstrap tree #345, logLikelihood: -45818.834600\n",
      "[03:22:00] [worker #1] Bootstrap tree #344, logLikelihood: -45103.901750\n",
      "[03:22:54] [worker #0] Bootstrap tree #340, logLikelihood: -48193.986620\n",
      "[03:23:41] [worker #2] Bootstrap tree #348, logLikelihood: -45923.782536\n",
      "[03:23:54] [worker #1] Bootstrap tree #347, logLikelihood: -45901.074617\n",
      "[03:24:23] [worker #0] Bootstrap tree #343, logLikelihood: -46409.302135\n",
      "[03:25:22] [worker #1] Bootstrap tree #350, logLikelihood: -44023.855421\n",
      "[03:25:45] [worker #0] Bootstrap tree #346, logLikelihood: -45394.117227\n",
      "[03:26:59] [worker #0] Bootstrap tree #349, logLikelihood: -46184.381968\n",
      "[03:28:39] [worker #2] Bootstrap tree #351, logLikelihood: -45634.639135\n",
      "[03:29:01] [worker #1] Bootstrap tree #353, logLikelihood: -45674.811987\n",
      "[03:29:01] [worker #0] Bootstrap tree #352, logLikelihood: -42347.560909\n",
      "[03:30:12] [worker #2] Bootstrap tree #354, logLikelihood: -46141.502680\n",
      "[03:30:28] [worker #1] Bootstrap tree #356, logLikelihood: -43972.301715\n",
      "[03:30:33] [worker #0] Bootstrap tree #355, logLikelihood: -45425.533981\n",
      "[03:31:46] [worker #2] Bootstrap tree #357, logLikelihood: -48201.454213\n",
      "[03:31:55] [worker #1] Bootstrap tree #359, logLikelihood: -42010.887916\n",
      "[03:32:06] [worker #0] Bootstrap tree #358, logLikelihood: -46701.920875\n",
      "[03:33:12] [worker #2] Bootstrap tree #360, logLikelihood: -46322.245659\n",
      "[03:33:16] [worker #1] Bootstrap tree #362, logLikelihood: -45695.529464\n",
      "[03:33:55] [worker #0] Bootstrap tree #361, logLikelihood: -46560.728344\n",
      "[03:34:45] [worker #1] Bootstrap tree #365, logLikelihood: -45382.925828\n",
      "[03:35:03] [worker #2] Bootstrap tree #363, logLikelihood: -42833.179416\n",
      "[03:36:13] [worker #0] Bootstrap tree #364, logLikelihood: -46099.299520\n",
      "[03:36:14] [worker #1] Bootstrap tree #368, logLikelihood: -45979.593330\n",
      "[03:36:33] [worker #2] Bootstrap tree #366, logLikelihood: -45141.285902\n",
      "[03:37:39] [worker #0] Bootstrap tree #367, logLikelihood: -46867.533028\n",
      "[03:37:49] [worker #1] Bootstrap tree #371, logLikelihood: -46512.199567\n",
      "[03:38:03] [worker #2] Bootstrap tree #369, logLikelihood: -45804.578312\n",
      "[03:39:22] [worker #0] Bootstrap tree #370, logLikelihood: -44058.935374\n",
      "[03:39:36] [worker #1] Bootstrap tree #374, logLikelihood: -46700.773233\n",
      "[03:39:52] [worker #2] Bootstrap tree #372, logLikelihood: -43934.127771\n",
      "[03:40:55] [worker #0] Bootstrap tree #373, logLikelihood: -46204.093529\n",
      "[03:40:55] [worker #1] Bootstrap tree #377, logLikelihood: -47343.414985\n",
      "[03:41:38] [worker #2] Bootstrap tree #375, logLikelihood: -46991.033265\n",
      "[03:42:20] [worker #1] Bootstrap tree #380, logLikelihood: -44077.060330\n",
      "[03:42:45] [worker #0] Bootstrap tree #376, logLikelihood: -44567.536690\n",
      "[03:43:03] [worker #2] Bootstrap tree #378, logLikelihood: -43420.702399\n",
      "[03:43:55] [worker #1] Bootstrap tree #383, logLikelihood: -46525.666595\n",
      "[03:44:24] [worker #0] Bootstrap tree #379, logLikelihood: -45698.319796\n",
      "[03:44:27] [worker #2] Bootstrap tree #381, logLikelihood: -43306.135707\n",
      "[03:45:22] [worker #1] Bootstrap tree #386, logLikelihood: -45318.943867\n",
      "[03:46:08] [worker #0] Bootstrap tree #382, logLikelihood: -44276.520350\n",
      "[03:46:14] [worker #2] Bootstrap tree #384, logLikelihood: -46118.567649\n",
      "[03:46:42] [worker #1] Bootstrap tree #389, logLikelihood: -45023.374370\n",
      "[03:47:31] [worker #2] Bootstrap tree #387, logLikelihood: -44398.148655\n",
      "[03:47:32] [worker #0] Bootstrap tree #385, logLikelihood: -45184.943536\n",
      "[03:48:05] [worker #1] Bootstrap tree #392, logLikelihood: -45582.531773\n",
      "[03:48:54] [worker #2] Bootstrap tree #390, logLikelihood: -44327.638738\n",
      "[03:49:02] [worker #0] Bootstrap tree #388, logLikelihood: -45418.903333\n",
      "[03:49:28] [worker #1] Bootstrap tree #395, logLikelihood: -45301.680483\n",
      "[03:50:27] [worker #2] Bootstrap tree #393, logLikelihood: -45506.517498\n",
      "[03:50:38] [worker #0] Bootstrap tree #391, logLikelihood: -43830.297405\n",
      "[03:51:23] [worker #1] Bootstrap tree #398, logLikelihood: -45584.310022\n",
      "[03:51:57] [worker #0] Bootstrap tree #394, logLikelihood: -42617.294599\n",
      "[03:51:58] [worker #2] Bootstrap tree #396, logLikelihood: -44020.720766\n",
      "[03:53:18] [worker #2] Bootstrap tree #399, logLikelihood: -43365.933664\n",
      "[03:53:21] [worker #0] Bootstrap tree #397, logLikelihood: -44435.880727\n",
      "[03:54:45] [worker #0] Bootstrap tree #400, logLikelihood: -43526.364070\n",
      "[03:56:13] [worker #0] Bootstrap tree #403, logLikelihood: -45656.602073\n",
      "[03:56:16] [worker #1] Bootstrap tree #401, logLikelihood: -43357.752279\n",
      "[03:56:41] [worker #2] Bootstrap tree #402, logLikelihood: -46476.002284\n",
      "[03:57:34] [worker #1] Bootstrap tree #404, logLikelihood: -46197.177016\n",
      "[03:57:45] [worker #0] Bootstrap tree #406, logLikelihood: -46141.104815\n",
      "[03:58:04] [worker #2] Bootstrap tree #405, logLikelihood: -44486.688680\n",
      "[03:59:07] [worker #1] Bootstrap tree #407, logLikelihood: -45264.043701\n",
      "[03:59:16] [worker #0] Bootstrap tree #409, logLikelihood: -45395.559102\n",
      "[03:59:31] [worker #2] Bootstrap tree #408, logLikelihood: -45889.253552\n",
      "[04:00:36] [worker #1] Bootstrap tree #410, logLikelihood: -45846.465373\n",
      "[04:00:42] [worker #0] Bootstrap tree #412, logLikelihood: -46730.742494\n",
      "[04:00:57] [worker #2] Bootstrap tree #411, logLikelihood: -44352.817512\n",
      "[04:02:10] [worker #1] Bootstrap tree #413, logLikelihood: -50114.487817\n",
      "[04:02:25] [worker #2] Bootstrap tree #414, logLikelihood: -46492.944323\n",
      "[04:02:37] [worker #0] Bootstrap tree #415, logLikelihood: -43176.847691\n",
      "[04:03:39] [worker #1] Bootstrap tree #416, logLikelihood: -46014.258841\n",
      "[04:03:49] [worker #2] Bootstrap tree #417, logLikelihood: -45645.229377\n",
      "[04:04:02] [worker #0] Bootstrap tree #418, logLikelihood: -43713.521249\n",
      "[04:05:07] [worker #1] Bootstrap tree #419, logLikelihood: -44983.673184\n",
      "[04:05:12] [worker #2] Bootstrap tree #420, logLikelihood: -41906.431849\n",
      "[04:06:12] [worker #0] Bootstrap tree #421, logLikelihood: -43556.286822\n",
      "[04:06:30] [worker #1] Bootstrap tree #422, logLikelihood: -42863.739108\n",
      "[04:06:38] [worker #2] Bootstrap tree #423, logLikelihood: -45623.255960\n",
      "[04:07:39] [worker #0] Bootstrap tree #424, logLikelihood: -44659.584302\n",
      "[04:08:09] [worker #2] Bootstrap tree #426, logLikelihood: -44875.359595\n",
      "[04:08:29] [worker #1] Bootstrap tree #425, logLikelihood: -43028.335556\n",
      "[04:09:11] [worker #0] Bootstrap tree #427, logLikelihood: -45817.752536\n",
      "[04:09:52] [worker #1] Bootstrap tree #428, logLikelihood: -43833.289691\n",
      "[04:10:46] [worker #2] Bootstrap tree #429, logLikelihood: -45021.752425\n",
      "[04:10:58] [worker #0] Bootstrap tree #430, logLikelihood: -45361.835429\n",
      "[04:11:12] [worker #1] Bootstrap tree #431, logLikelihood: -43397.222122\n",
      "[04:12:08] [worker #2] Bootstrap tree #432, logLikelihood: -44545.711466\n",
      "[04:12:35] [worker #1] Bootstrap tree #434, logLikelihood: -45586.851238\n",
      "[04:12:44] [worker #0] Bootstrap tree #433, logLikelihood: -43789.801139\n",
      "[04:13:43] [worker #2] Bootstrap tree #435, logLikelihood: -46256.488713\n",
      "[04:14:10] [worker #0] Bootstrap tree #436, logLikelihood: -46584.482467\n",
      "[04:14:22] [worker #1] Bootstrap tree #437, logLikelihood: -44289.545750\n",
      "[04:15:28] [worker #2] Bootstrap tree #438, logLikelihood: -44215.559672\n",
      "[04:15:37] [worker #0] Bootstrap tree #439, logLikelihood: -44839.132783\n",
      "[04:15:56] [worker #1] Bootstrap tree #440, logLikelihood: -47097.021720\n",
      "[04:17:06] [worker #0] Bootstrap tree #442, logLikelihood: -45787.566049\n",
      "[04:17:23] [worker #1] Bootstrap tree #443, logLikelihood: -43481.632685\n",
      "[04:17:26] [worker #2] Bootstrap tree #441, logLikelihood: -46115.791576\n",
      "[04:18:30] [worker #0] Bootstrap tree #445, logLikelihood: -45895.384900\n",
      "[04:18:54] [worker #2] Bootstrap tree #444, logLikelihood: -45195.343282\n",
      "[04:18:55] [worker #1] Bootstrap tree #446, logLikelihood: -44653.910931\n",
      "[04:20:41] [worker #0] Bootstrap tree #448, logLikelihood: -45474.356319\n",
      "[04:20:54] [worker #2] Bootstrap tree #447, logLikelihood: -47215.416594\n",
      "[04:21:06] [worker #1] Bootstrap tree #449, logLikelihood: -46015.451266\n",
      "[04:22:15] [worker #2] Bootstrap tree #450, logLikelihood: -45412.614855\n",
      "[04:23:45] [worker #0] Bootstrap tree #451, logLikelihood: -45452.613561\n",
      "[04:23:46] [worker #1] Bootstrap tree #452, logLikelihood: -43987.932244\n",
      "[04:23:55] [worker #2] Bootstrap tree #453, logLikelihood: -46349.043295\n",
      "[04:25:16] [worker #1] Bootstrap tree #455, logLikelihood: -45358.474460\n",
      "[04:25:20] [worker #0] Bootstrap tree #454, logLikelihood: -45605.603078\n",
      "[04:25:34] [worker #2] Bootstrap tree #456, logLikelihood: -46949.432077\n",
      "[04:26:59] [worker #1] Bootstrap tree #458, logLikelihood: -44098.996220\n",
      "[04:27:05] [worker #2] Bootstrap tree #459, logLikelihood: -41903.460106\n",
      "[04:27:11] [worker #0] Bootstrap tree #457, logLikelihood: -46811.401085\n",
      "[04:28:47] [worker #0] Bootstrap tree #460, logLikelihood: -45762.383579\n",
      "[04:29:11] [worker #1] Bootstrap tree #461, logLikelihood: -46176.189910\n",
      "[04:29:23] [worker #2] Bootstrap tree #462, logLikelihood: -46835.098078\n",
      "[04:30:34] [worker #1] Bootstrap tree #464, logLikelihood: -43518.768330\n",
      "[04:31:04] [worker #0] Bootstrap tree #463, logLikelihood: -47630.958005\n",
      "[04:31:20] [worker #2] Bootstrap tree #465, logLikelihood: -45663.607902\n",
      "[04:32:10] [worker #1] Bootstrap tree #467, logLikelihood: -45251.979361\n",
      "[04:32:50] [worker #2] Bootstrap tree #468, logLikelihood: -44084.932237\n",
      "[04:33:50] [worker #1] Bootstrap tree #470, logLikelihood: -46492.233451\n",
      "[04:33:59] [worker #0] Bootstrap tree #466, logLikelihood: -45842.683788\n",
      "[04:34:23] [worker #2] Bootstrap tree #471, logLikelihood: -45085.239295\n",
      "[04:35:30] [worker #1] Bootstrap tree #473, logLikelihood: -45304.034713\n",
      "[04:35:36] [worker #0] Bootstrap tree #469, logLikelihood: -45341.701832\n",
      "[04:35:50] [worker #2] Bootstrap tree #474, logLikelihood: -45262.038404\n",
      "[04:37:01] [worker #1] Bootstrap tree #476, logLikelihood: -45382.297439\n",
      "[04:37:15] [worker #0] Bootstrap tree #472, logLikelihood: -43237.725128\n",
      "[04:37:35] [worker #2] Bootstrap tree #477, logLikelihood: -45876.888763\n",
      "[04:38:41] [worker #1] Bootstrap tree #479, logLikelihood: -44436.291651\n",
      "[04:38:52] [worker #0] Bootstrap tree #475, logLikelihood: -45626.263961\n",
      "[04:39:03] [worker #2] Bootstrap tree #480, logLikelihood: -44806.071226\n",
      "[04:40:08] [worker #1] Bootstrap tree #482, logLikelihood: -45578.041554\n",
      "[04:40:33] [worker #0] Bootstrap tree #478, logLikelihood: -43601.503692\n",
      "[04:40:42] [worker #2] Bootstrap tree #483, logLikelihood: -47453.823904\n",
      "[04:42:00] [worker #0] Bootstrap tree #481, logLikelihood: -45816.550450\n",
      "[04:42:18] [worker #1] Bootstrap tree #485, logLikelihood: -46114.171024\n",
      "[04:42:19] [worker #2] Bootstrap tree #486, logLikelihood: -43724.341030\n",
      "[04:43:47] [worker #0] Bootstrap tree #484, logLikelihood: -47118.521013\n",
      "[04:43:53] [worker #1] Bootstrap tree #488, logLikelihood: -45685.412193\n",
      "[04:44:10] [worker #2] Bootstrap tree #489, logLikelihood: -45874.450481\n",
      "[04:45:26] [worker #1] Bootstrap tree #491, logLikelihood: -42914.644960\n",
      "[04:45:33] [worker #0] Bootstrap tree #487, logLikelihood: -45056.511565\n",
      "[04:45:39] [worker #2] Bootstrap tree #492, logLikelihood: -43914.476176\n",
      "[04:47:05] [worker #1] Bootstrap tree #494, logLikelihood: -44958.768717\n",
      "[04:47:08] [worker #0] Bootstrap tree #490, logLikelihood: -43791.336729\n",
      "[04:47:19] [worker #2] Bootstrap tree #495, logLikelihood: -45636.585921\n",
      "[04:48:35] [worker #0] Bootstrap tree #493, logLikelihood: -45518.972877\n",
      "[04:48:40] [worker #1] Bootstrap tree #497, logLikelihood: -45746.856286\n",
      "[04:49:00] [worker #2] Bootstrap tree #498, logLikelihood: -45294.939608\n",
      "[04:50:02] [worker #0] Bootstrap tree #496, logLikelihood: -46244.276031\n",
      "[04:50:15] [worker #1] Bootstrap tree #500, logLikelihood: -45053.533154\n",
      "[04:51:35] [worker #0] Bootstrap tree #499, logLikelihood: -45912.320125\n",
      "[04:53:05] [worker #2] Bootstrap tree #501, logLikelihood: -46897.993857\n",
      "[04:53:15] [worker #0] Bootstrap tree #502, logLikelihood: -44078.347034\n",
      "[04:53:18] [worker #1] Bootstrap tree #503, logLikelihood: -45676.934819\n",
      "[04:54:41] [worker #2] Bootstrap tree #504, logLikelihood: -45292.110937\n",
      "[04:54:50] [worker #0] Bootstrap tree #505, logLikelihood: -43797.474835\n",
      "[04:54:54] [worker #1] Bootstrap tree #506, logLikelihood: -43426.130216\n",
      "[04:56:20] [worker #0] Bootstrap tree #508, logLikelihood: -46913.862312\n",
      "[04:56:36] [worker #1] Bootstrap tree #509, logLikelihood: -43106.465323\n",
      "[04:56:38] [worker #2] Bootstrap tree #507, logLikelihood: -45660.944306\n",
      "[04:57:54] [worker #0] Bootstrap tree #511, logLikelihood: -44834.137279\n",
      "[04:58:16] [worker #1] Bootstrap tree #512, logLikelihood: -45470.629176\n",
      "[04:58:30] [worker #2] Bootstrap tree #510, logLikelihood: -45515.779747\n",
      "[04:59:33] [worker #0] Bootstrap tree #514, logLikelihood: -45410.317185\n",
      "[05:00:03] [worker #2] Bootstrap tree #513, logLikelihood: -45872.484779\n",
      "[05:00:11] [worker #1] Bootstrap tree #515, logLikelihood: -43526.542446\n",
      "[05:01:11] [worker #0] Bootstrap tree #517, logLikelihood: -45400.229343\n",
      "[05:01:49] [worker #1] Bootstrap tree #518, logLikelihood: -44655.982438\n",
      "[05:01:56] [worker #2] Bootstrap tree #516, logLikelihood: -44255.046489\n",
      "[05:03:02] [worker #0] Bootstrap tree #520, logLikelihood: -45527.101933\n",
      "[05:03:15] [worker #1] Bootstrap tree #521, logLikelihood: -43595.442575\n",
      "[05:03:36] [worker #2] Bootstrap tree #519, logLikelihood: -44088.123981\n",
      "[05:04:27] [worker #0] Bootstrap tree #523, logLikelihood: -43923.082034\n",
      "[05:04:46] [worker #1] Bootstrap tree #524, logLikelihood: -46675.658920\n",
      "[05:05:23] [worker #2] Bootstrap tree #522, logLikelihood: -47271.070307\n",
      "[05:06:15] [worker #0] Bootstrap tree #526, logLikelihood: -43581.116026\n",
      "[05:06:20] [worker #1] Bootstrap tree #527, logLikelihood: -46289.653810\n",
      "[05:07:11] [worker #2] Bootstrap tree #525, logLikelihood: -46218.219268\n",
      "[05:07:50] [worker #1] Bootstrap tree #530, logLikelihood: -44116.796029\n",
      "[05:08:11] [worker #0] Bootstrap tree #529, logLikelihood: -45277.363990\n",
      "[05:08:35] [worker #2] Bootstrap tree #528, logLikelihood: -45988.332504\n",
      "[05:09:40] [worker #1] Bootstrap tree #533, logLikelihood: -44311.541099\n",
      "[05:09:51] [worker #0] Bootstrap tree #532, logLikelihood: -47103.413650\n",
      "[05:10:23] [worker #2] Bootstrap tree #531, logLikelihood: -47793.289351\n",
      "[05:11:23] [worker #0] Bootstrap tree #535, logLikelihood: -44985.446232\n",
      "[05:11:53] [worker #1] Bootstrap tree #536, logLikelihood: -44144.862314\n",
      "[05:12:15] [worker #2] Bootstrap tree #534, logLikelihood: -46515.774682\n",
      "[05:13:01] [worker #0] Bootstrap tree #538, logLikelihood: -46321.454987\n",
      "[05:13:23] [worker #1] Bootstrap tree #539, logLikelihood: -43916.602152\n",
      "[05:14:04] [worker #2] Bootstrap tree #537, logLikelihood: -47220.268032\n",
      "[05:14:46] [worker #0] Bootstrap tree #541, logLikelihood: -45092.606203\n",
      "[05:14:48] [worker #1] Bootstrap tree #542, logLikelihood: -43573.814152\n",
      "[05:15:22] [worker #2] Bootstrap tree #540, logLikelihood: -44802.982154\n",
      "[05:16:19] [worker #1] Bootstrap tree #545, logLikelihood: -42663.785127\n",
      "[05:16:41] [worker #0] Bootstrap tree #544, logLikelihood: -48162.545777\n",
      "[05:16:53] [worker #2] Bootstrap tree #543, logLikelihood: -47415.569602\n",
      "[05:17:47] [worker #1] Bootstrap tree #548, logLikelihood: -45338.054415\n",
      "[05:18:17] [worker #0] Bootstrap tree #547, logLikelihood: -45422.503257\n",
      "[05:18:27] [worker #2] Bootstrap tree #546, logLikelihood: -43558.135317\n",
      "[05:19:48] [worker #2] Bootstrap tree #549, logLikelihood: -44372.310360\n",
      "[05:20:01] [worker #0] Bootstrap tree #550, logLikelihood: -45244.973659\n",
      "[05:20:03] Bootstrapping converged after 550 replicates.\n",
      "\n",
      "Bootstrap trees saved to: /home/lubo/code/wRajter/vampyrella_2023/notebooks/T3_bootstraps.raxml.bootstraps\n",
      "\n",
      "Execution log saved to: /home/lubo/code/wRajter/vampyrella_2023/notebooks/T3_bootstraps.raxml.log\n",
      "\n",
      "Analysis started: 11-Dec-2023 12:33:00 / finished: 11-Dec-2023 17:53:04\n",
      "\n",
      "Elapsed time: 19204.127 seconds\n",
      "\n",
      "\n",
      "Output files were moved to: ../results/phyl_trees/all_projects_combined/T3_bootstraps\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prefix = 'T3_bootstraps'\n",
    "model = 'GTR+G'\n",
    "# output_dir = os.path.join(results_path, 'phyl_trees', 'refrence_alignment_tree', prefix)\n",
    "output_dir = os.path.join(results_path, 'phyl_trees', 'all_projects_combined', prefix)\n",
    "\n",
    "# Check if the output directory exists, create it if not\n",
    "if not os.path.exists(output_dir):\n",
    "    os.makedirs(output_dir)\n",
    "\n",
    "\n",
    "run_raxmlng_bootstrap(alignment = alignment,\n",
    "                      output_dir = output_dir, \n",
    "                      model = model, \n",
    "                      prefix = prefix)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5d40766-3578-4e56-bd33-0b948f104cd0",
   "metadata": {},
   "source": [
    "### Computing branch support"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a608ba9e-2a97-41e5-a027-2fb777382891",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "RAxML-NG v. 1.1.0 released on 29.11.2021 by The Exelixis Lab.\n",
      "Developed by: Alexey M. Kozlov and Alexandros Stamatakis.\n",
      "Contributors: Diego Darriba, Tomas Flouri, Benoit Morel, Sarah Lutteropp, Ben Bettisworth.\n",
      "Latest version: https://github.com/amkozlov/raxml-ng\n",
      "Questions/problems/suggestions? Please visit: https://groups.google.com/forum/#!forum/raxml\n",
      "\n",
      "System: 11th Gen Intel(R) Core(TM) i7-11850H @ 2.50GHz, 8 cores, 11 GB RAM\n",
      "\n",
      "RAxML-NG was called at 11-Dec-2023 19:03:02 as follows:\n",
      "\n",
      "raxml-ng --support --tree ../results/phyl_trees/all_projects_combined/T2_MLtree/T2_MLtree.raxml.bestTree --bs-trees ../results/phyl_trees/all_projects_combined/T3_bootstraps/T3_bootstraps.raxml.bootstraps --prefix T4_supports --threads 3\n",
      "\n",
      "Analysis options:\n",
      "  run mode: Compute bipartition support (Felsenstein Bootstrap)\n",
      "  start tree(s): user\n",
      "  random seed: 1702317782\n",
      "  SIMD kernels: AVX2\n",
      "  parallelization: coarse-grained (auto), PTHREADS (3 threads), thread pinning: OFF\n",
      "\n",
      "Reading reference tree from file: ../results/phyl_trees/all_projects_combined/T2_MLtree/T2_MLtree.raxml.bestTree\n",
      "Reference tree size: 183\n",
      "\n",
      "Reading bootstrap trees from file: ../results/phyl_trees/all_projects_combined/T3_bootstraps/T3_bootstraps.raxml.bootstraps\n",
      "Loaded 550 trees with 183 taxa.\n",
      "\n",
      "Best ML tree with Felsenstein bootstrap (FBP) support values saved to: /home/lubo/code/wRajter/vampyrella_2023/notebooks/T4_supports.raxml.support\n",
      "\n",
      "Execution log saved to: /home/lubo/code/wRajter/vampyrella_2023/notebooks/T4_supports.raxml.log\n",
      "\n",
      "Analysis started: 11-Dec-2023 19:03:02 / finished: 11-Dec-2023 19:03:02\n",
      "\n",
      "Elapsed time: 0.228 seconds\n",
      "\n",
      "\n",
      "Output files were moved to: ../results/phyl_trees/all_projects_combined/T4_supports\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# main_dir = os.path.join(results_path, 'phyl_trees', 'refrence_alignment_tree')\n",
    "\n",
    "main_dir = os.path.join(results_path, 'phyl_trees', 'all_projects_combined')\n",
    "\n",
    "bootstraps = os.path.join(main_dir, 'T3_bootstraps', 'T3_bootstraps.raxml.bootstraps')\n",
    "tree = os.path.join(main_dir, 'T2_MLtree', 'T2_MLtree.raxml.bestTree')\n",
    "prefix = 'T4_supports'\n",
    "output_dir = os.path.join(main_dir, prefix)\n",
    "\n",
    "# Check if the output directory exists, create it if not\n",
    "if not os.path.exists(output_dir):\n",
    "    os.makedirs(output_dir)\n",
    "\n",
    "run_raxmlng_support(tree = tree,\n",
    "                    output_dir = output_dir, \n",
    "                    bootstraps = bootstraps, \n",
    "                    num_threads=3, \n",
    "                    prefix=prefix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d04e30e-60be-4bbf-b33d-7ea2fa32f631",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "7ddef389-710d-4242-a6da-e5e4652e3ace",
   "metadata": {},
   "source": [
    "# Tree visualisation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea38adb5-e453-4d15-913d-8001b5dd3ffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "803a8b00-256d-4d92-ab0e-e29a24ffd259",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Variables\n",
    "tree_path = os.path.join(results_path, 'phyl_trees', project, marker, denoise_method, sim, 'T2_gblocks_strong', 'T2.raxml.bestTree')\n",
    "alignment_path = os.path.join(results_path, 'phyl_trees', project, marker, denoise_method, sim, 'alignments', 'vamp_mafft.fasta')\n",
    "os.environ['QT_QPA_PLATFORM'] = 'offscreen'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d99bbf1f-5e33-420a-a1ce-873d82fa69cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating name_mapping dictionary to rename the new sequences in the tree\n",
    "\n",
    "name_mapping = {}\n",
    "\n",
    "with open(alignment_path, 'r') as f:\n",
    "    for line in f:\n",
    "        # Reference sequences part\n",
    "        if line.startswith('>') and 'Vampyrellida' in line:\n",
    "            old_name = line.strip('>').strip()\n",
    "            new_name = line.split('|')[-1].strip()\n",
    "            id = line.split('|')[0].strip('>').strip()\n",
    "            name_mapping[old_name] = new_name + '_' + id\n",
    "        # New environmental sequences part\n",
    "        elif line.startswith('>') and 'centroid' in line:\n",
    "            old_name = line.strip('>').strip()\n",
    "            seq_name = line.split('=')[1].strip('_seqs')\n",
    "            sample_name = line.split('_')[-1].strip()\n",
    "            new_name = 'env_otu_' + seq_name + '_' + sample_name\n",
    "            name_mapping[old_name] = new_name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1927fed0-1450-40a4-ad1c-8e92c54eb945",
   "metadata": {},
   "source": [
    "### Create inner nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a439ccda-7cd4-49e3-b85f-cc12d2d253ff",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Load the tree\n",
    "tree = Tree(tree_path)\n",
    "\n",
    "# Create a tree style\n",
    "ts = TreeStyle()\n",
    "\n",
    "# Styling\n",
    "ts.scale = 2000\n",
    "ts.tree_width = 60000  # Increase this for a wider tree\n",
    "\n",
    "# Assign names to inner nodes for easy reference\n",
    "i = 0\n",
    "for n in tree.traverse():\n",
    "    if not n.is_leaf():  # Check if it is an inner node\n",
    "        n.name = \"internal_node_{}\".format(i)\n",
    "        i += 1\n",
    "\n",
    "# for node in tree.traverse():\n",
    "#     print(node.name)\n",
    "\n",
    "# # Apply the renaming directly to the leaves of the tree\n",
    "# for leaf in tree.iter_leaves():\n",
    "#     if leaf.name in name_mapping:\n",
    "#         leaf.name = name_mapping[leaf.name]\n",
    "\n",
    "# Styling\n",
    "ts.show_leaf_name = True # display leaf names\n",
    "\n",
    "# Hides the node circles\n",
    "for node in tree.traverse():\n",
    "    node.img_style[\"size\"] = 0\n",
    "\n",
    "# Add node names/IDs\n",
    "for n in tree.traverse():\n",
    "    # Create a face to display node name/ID\n",
    "    name_face = TextFace(n.name, fsize=10, fgcolor='black')\n",
    "    n.add_face(name_face, column=0, position=\"branch-top\")\n",
    "\n",
    "tree.render(os.path.join(results_path, 'phyl_trees', project, marker, denoise_method, sim, 'tree_inner_nodes_gblocks_strong.png'), dpi=300, w=10000, tree_style=ts)\n",
    "tree.render('%%inline', tree_style=ts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b01dba8-ea00-4880-a349-4999f78dfcc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the tree\n",
    "tree = Tree(tree_path)\n",
    "\n",
    "# Create a tree style\n",
    "ts = TreeStyle()\n",
    "# Styling\n",
    "ts.scale = 2000\n",
    "ts.tree_width = 60000  # Increase this for a wider tree\n",
    "\n",
    "# Assign names to inner nodes for easy reference\n",
    "i = 0\n",
    "for n in tree.traverse():\n",
    "    if not n.is_leaf():  # Check if it is an inner node\n",
    "        n.name = 'internal_node_{}'.format(i)\n",
    "        i += 1\n",
    "\n",
    "\n",
    "# Find the node you want to use as the new root\n",
    "node_to_reroot = tree&'internal_node_20'  # Replace 'Node_Name' with the name of your desired node\n",
    "\n",
    "# Use the assigned name to reroot\n",
    "tree.set_outgroup(node_to_reroot)\n",
    "\n",
    "# # Identify the node whose children you want to swap\n",
    "node_to_swap_1 = tree&'internal_node_0'  \n",
    "node_to_swap_2 = tree&'internal_node_12'\n",
    "# node_to_swap_3 = tree&'internal_node_68'\n",
    "# node_to_swap_4 = tree&'internal_node_64'\n",
    "# node_to_swap_5 = tree&'internal_node_72'\n",
    "\n",
    "# # Swap the children of the target node\n",
    "node_to_swap_1.swap_children()\n",
    "node_to_swap_2.swap_children()\n",
    "# node_to_swap_3.swap_children()\n",
    "# node_to_swap_4.swap_children()\n",
    "# node_to_swap_5.swap_children()\n",
    "\n",
    "\n",
    "\n",
    "# Apply the renaming directly to the leaves of the tree\n",
    "for leaf in tree.iter_leaves():\n",
    "    if leaf.name in name_mapping:\n",
    "        leaf.name = name_mapping[leaf.name]\n",
    "\n",
    "\n",
    "# Add node names/IDs\n",
    "for n in tree.traverse():\n",
    "    # Create a face to display node name/ID\n",
    "    name_face = TextFace(n.name, fsize=10, fgcolor='black')\n",
    "    n.add_face(name_face, column=0, position=\"branch-top\")\n",
    "\n",
    "\n",
    "# Hides the node circles\n",
    "for node in tree.traverse():\n",
    "    node.img_style['size'] = 0\n",
    "\n",
    "tree.render('%%inline', tree_style=ts)\n",
    "# tree.show()\n",
    "# tree.render(os.path.join(results_path, 'tree_test.png'), tree_style=ts)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f59c8d25-97ce-4e46-ba42-7f3050115ca4",
   "metadata": {},
   "source": [
    "### Topology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f02b72a5-33c1-4830-bafa-cb9757c76943",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the tree\n",
    "tree = Tree(tree_path)\n",
    "\n",
    "# Create a tree style\n",
    "ts = TreeStyle()\n",
    "# Styling\n",
    "ts.scale = 2000\n",
    "ts.tree_width = 60000  # Increase this for a wider tree\n",
    "line_width = 2\n",
    "\n",
    "\n",
    "# Assign names to inner nodes for easy reference\n",
    "i = 0\n",
    "for n in tree.traverse():\n",
    "    if not n.is_leaf():  # Check if it is an inner node\n",
    "        n.name = 'internal_node_{}'.format(i)\n",
    "        i += 1\n",
    "\n",
    "\n",
    "# Find the node you want to use as the new root\n",
    "node_to_reroot = tree&'internal_node_6'  # Replace 'Node_Name' with the name of your desired node\n",
    "\n",
    "# Use the assigned name to reroot\n",
    "tree.set_outgroup(node_to_reroot)\n",
    "\n",
    "# Identify the node whose children you want to swap\n",
    "node_to_swap_1 = tree&'internal_node_0'  \n",
    "node_to_swap_2 = tree&'internal_node_2'\n",
    "# node_to_swap_3 = tree&'internal_node_11'\n",
    "# node_to_swap_4 = tree&'internal_node_64'\n",
    "# node_to_swap_5 = tree&'internal_node_72'\n",
    "\n",
    "# Swap the children of the target node\n",
    "node_to_swap_1.swap_children()\n",
    "node_to_swap_2.swap_children()\n",
    "node_to_swap_3.swap_children()\n",
    "# node_to_swap_4.swap_children()\n",
    "# node_to_swap_5.swap_children()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Apply the renaming directly to the leaves of the tree\n",
    "for leaf in tree.iter_leaves():\n",
    "    if leaf.name in name_mapping:\n",
    "        leaf.name = name_mapping[leaf.name]\n",
    "\n",
    "\n",
    "# # Add node names/IDs\n",
    "# for n in tree.traverse():\n",
    "#     # Create a face to display node name/ID\n",
    "#     name_face = TextFace(n.name, fsize=10, fgcolor='black')\n",
    "#     n.add_face(name_face, column=0, position=\"branch-top\")\n",
    "\n",
    "\n",
    "# Hides the node circles\n",
    "for node in tree.traverse():\n",
    "    node.img_style['size'] = 0\n",
    "\n",
    "# Nodestyle\n",
    "for node in tree.traverse():\n",
    "    # Set the color and thickness for all nodes in the subtree\n",
    "    nstyle = NodeStyle()\n",
    "    nstyle[\"size\"] = 0  # Set size to 0 to hide the default circle node\n",
    "    nstyle[\"vt_line_width\"] = line_width  # Set the thickness of vertical lines\n",
    "    nstyle[\"hz_line_width\"] = line_width  # Set the thickness of horizontal lines\n",
    "    node.set_style(nstyle)\n",
    "\n",
    "# Find the node you want to collapse\n",
    "node_to_collapse_1 = tree&'internal_node_6'\n",
    "node_to_collapse_2 = tree&'internal_node_7'\n",
    "node_to_collapse_3 = tree&'internal_node_29'\n",
    "node_to_collapse_4 = tree&'internal_node_28'\n",
    "node_to_collapse_5 = tree&'internal_node_1'\n",
    "node_to_collapse_6 = tree&'internal_node_8'\n",
    "node_to_collapse_7 = tree&'internal_node_19'\n",
    "\n",
    "# node_to_collapse.img_style['shape'] = 'sphere'\n",
    "# node_to_collapse.img_style['size'] = 50\n",
    "node_to_collapse_1.img_style['draw_descendants'] = False\n",
    "node_to_collapse_2.img_style['draw_descendants'] = False\n",
    "node_to_collapse_3.img_style['draw_descendants'] = False\n",
    "node_to_collapse_4.img_style['draw_descendants'] = False\n",
    "node_to_collapse_5.img_style['draw_descendants'] = False\n",
    "node_to_collapse_6.img_style['draw_descendants'] = False\n",
    "node_to_collapse_7.img_style['draw_descendants'] = False\n",
    "\n",
    "# Rename the node\n",
    "node_to_collapse_1.name = 'Sericomyxidae'\n",
    "node_to_collapse_2.name = 'Thalassomyxa & Arachnula'\n",
    "node_to_collapse_3.name = 'Placopodidae'\n",
    "node_to_collapse_4.name = 'B4'\n",
    "node_to_collapse_5.name = 'Vampyrella & Leptophrys'\n",
    "node_to_collapse_6.name = 'B1'\n",
    "node_to_collapse_7.name = 'B2'\n",
    "\n",
    "tree.render('%%inline', tree_style=ts)\n",
    "# tree.show()\n",
    "tree.render(os.path.join(results_path, 'phyl_trees', project, marker, denoise_method, sim, 'tree_gblocks_strong.png'), dpi=600, w=1000, tree_style=ts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17fc34e7-1f98-42ef-a787-dda74caafbbb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4496e386-a893-46d8-bd01-d889f10df18c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ad65409-0a41-4aed-a89b-f252a57704bd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8eec7ac1-08ca-48c4-93ce-8f60f89ab642",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8e115f6-f1a9-442c-a2bc-a0d7f6f67d0e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d95d39f-7c14-4c5c-89a0-c19ae2e2138a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4819b325-b0d4-4e7e-9021-da810d43c293",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from ete3 import Tree, TreeStyle, NodeStyle, TextFace\n",
    "\n",
    "os.environ['QT_QPA_PLATFORM'] = 'offscreen'\n",
    "line_width = 1\n",
    "tree_width = 200\n",
    "\n",
    "\n",
    "# Create a Tree object from the Newick file\n",
    "main_tree = Tree(f'{tree_file}/T2.raxml.bestTree')\n",
    "\n",
    "\n",
    "# Apply the renaming directly to the leaves of the tree\n",
    "for leaf in main_tree.iter_leaves():\n",
    "    if leaf.name in name_mapping:\n",
    "        leaf.name = name_mapping[leaf.name]\n",
    "        \n",
    "\n",
    "# Function to get color based on the label\n",
    "def get_color(name):\n",
    "    if 'Leptophrys_vorax_strain_LV02_' in name:\n",
    "        return 'red'\n",
    "    elif 'Leptophrys_vorax_strain_LV04_' in name:\n",
    "        return '#990000'\n",
    "    elif 'Vampyrella_lateritia_strain_VL09_' in name:\n",
    "        return 'green'\n",
    "    elif 'Vampyrella_lateritia_strain_NHR1' in name:\n",
    "        return '#33CC33'\n",
    "    elif 'Pseudovampyrella_closterii_strain_VC01_' in name:\n",
    "        return 'purple'\n",
    "    elif 'Placopus_flabellus_NHR1_' in name:\n",
    "        return 'orange'\n",
    "    elif 'Leptophrys_vorax_strain_LV01_' in name:\n",
    "        return '#FF6666'\n",
    "    return 'black'  # Default color if no match\n",
    "\n",
    "for node in main_tree.traverse():\n",
    "    # Set the color and thickness for all nodes in the subtree\n",
    "    nstyle = NodeStyle()\n",
    "    nstyle[\"size\"] = 0  # Set size to 0 to hide the default circle node\n",
    "    nstyle[\"vt_line_width\"] = line_width  # Set the thickness of vertical lines\n",
    "    nstyle[\"hz_line_width\"] = line_width  # Set the thickness of horizontal lines\n",
    "    node.set_style(nstyle)\n",
    "\n",
    "for leaf in main_tree.iter_leaves():\n",
    "    color = get_color(leaf.name)\n",
    "\n",
    "    # Apply the color to the text (leaf name)\n",
    "    name_face = TextFace(leaf.name, fgcolor=color)\n",
    "    leaf.add_face(name_face, column=0, position=\"branch-right\")\n",
    "\n",
    "    # Apply the color to the branch leading to the leaf\n",
    "    nstyle = NodeStyle()\n",
    "    nstyle[\"fgcolor\"] = color\n",
    "    nstyle[\"size\"] = 0  # Set size to 0 to hide the default circle node\n",
    "    nstyle[\"vt_line_color\"] = color\n",
    "    nstyle[\"hz_line_color\"] = color\n",
    "    nstyle[\"vt_line_width\"] = line_width  # Control the thickness of the vertical lines\n",
    "    nstyle[\"hz_line_width\"] = line_width  # Control the thickness of the horizontal lines\n",
    "    leaf.set_style(nstyle)\n",
    "\n",
    "\n",
    "ts = TreeStyle()\n",
    "ts.scale = tree_width  # Increase this value to spread out the tree more\n",
    "ts.show_leaf_name = False\n",
    "# main_tree.render(\"%%inline\", tree_style=ts)\n",
    "# main_tree.render(f\"{tree_file}/tree_mock.png\", dpi=600, w=10000, tree_style=ts)\n",
    "\n",
    "# save as Newick\n",
    "# main_tree.write(outfile=f\"{tree_file}/tree_mock.newick\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0df2cf47-acee-4a49-9edf-c5882413ed98",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "63bd058c-0691-4ea9-9c7f-179aa35ce17b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Cola"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9191b9c0-321a-44ff-9b10-86f4c36c6daf",
   "metadata": {},
   "outputs": [],
   "source": [
    "from ete3 import Tree\n",
    "\n",
    "def collapse_low_support_nodes(tree_path, threshold=70):\n",
    "    \"\"\"\n",
    "    Collapse nodes in a phylogenetic tree that have bootstrap support values below the given threshold.\n",
    "    \n",
    "    Parameters:\n",
    "    tree_path (str): Path to the Newick tree file.\n",
    "    threshold (int): Bootstrap support threshold below which nodes will be collapsed. Defaults to 70.\n",
    "    \n",
    "    Returns:\n",
    "    Tree: The tree with nodes collapsed below the bootstrap support threshold.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Load the tree\n",
    "    tree = Tree(tree_path)\n",
    "    \n",
    "    # Iterate over each node\n",
    "    for node in tree.traverse():\n",
    "        # If the node has a bootstrap support value and it's below the threshold\n",
    "        if node.support < threshold:\n",
    "            # Collapse the node\n",
    "            node.delete()\n",
    "    \n",
    "    return tree\n",
    "\n",
    "# Example usage\n",
    "collapsed_tree = collapse_low_support_nodes('path/to/your/tree.nw', threshold=70)\n",
    "collapsed_tree.write(outfile='path/to/your/collapsed_tree.nw')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ec5e3d3-afd6-4dfc-a99d-56418b912e5f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dda219e7-1bf5-42b5-9d50-32db89073441",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56d29678-5882-437c-9e9c-3bfd62045141",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a614fc16-6d85-4b02-822b-decb8a56f4e4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3195f01b-072f-4703-8049-db9c87cabf16",
   "metadata": {},
   "source": [
    "# Subtrees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9aa856d-960b-4245-97c2-a90006a493bc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
