{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "12461c61",
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "import Bio\n",
    "from Bio import AlignIO\n",
    "from Bio.Align.Applications import ClustalwCommandline\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import subprocess\n",
    "from ete3 import Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "523065d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# variables\n",
    "project = 'Suthaus_2022'\n",
    "cell = 'cellCombined'\n",
    "marker = 'Full18S'\n",
    "sim = 'sim95'\n",
    "raw_data = '../raw_data'\n",
    "denoise_method = 'FAD'\n",
    "vamp_spec_dir = f'{raw_data}/vamp_specific_seqs/{project}/{marker}/{cell}/{sim}/{denoise_method}'\n",
    "vamp_ref_seqs = f'{raw_data}/reference_alignments/vamp/2023_VAMPYRELLIDA_SSU_annotated_PR2_AM_AS_not_aligned_version.fasta'\n",
    "clustalw_path = f'{raw_data}/packages/clustalw-2.1-linux-x86_64-libcppstatic/clustalw2'\n",
    "tree_file = f'{raw_data}/phylogenetic_inference/{project}/{marker}/{cell}/{sim}'\n",
    "pics_dir = f'{raw_data}/OTU_results/{project}'\n",
    "tax_assign_dir = f'{raw_data}/tax_assign_results/{project}/{marker}/{cell}/{sim}/{denoise_method}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a541af8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# functions\n",
    "\n",
    "def strip(file_path, output_path='out_stripped.fasta'):\n",
    "    '''\n",
    "    Remove line breaks in FASTA files\n",
    "    '''\n",
    "    seqs = []\n",
    "    # read a file and stripped it\n",
    "    with open(file_path,'rt') as f:\n",
    "        lines = f.readlines()\n",
    "        for line in lines:\n",
    "            if line.startswith('>'):\n",
    "                seqs.append('\\n' + line)\n",
    "            else:\n",
    "                line_cleaned = line.rstrip()\n",
    "                seqs.append(line_cleaned)\n",
    "        seqs[0] = seqs[0].lstrip()\n",
    "    # write the file into fasta\n",
    "    with open(output_path, 'w') as fp:\n",
    "        for seq in seqs:\n",
    "            fp.write(seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e775effb",
   "metadata": {},
   "source": [
    "# Creating alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbc82c8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# format the ref. sequences fasta file if needed\n",
    "\n",
    "strip(file_path = vamp_ref_seqs, output_path = vamp_ref_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6de676a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine the ref. sequences fasta file and denoised sequences\n",
    "\n",
    "\n",
    "# Path to the output file\n",
    "file_combined_path = f'{vamp_spec_dir}/reference_and_mock_seqs.fasta'\n",
    "\n",
    "# Path to the denoised sequences\n",
    "denoised_sequences = f'{vamp_spec_dir}/Mock_18S_otu.fasta'\n",
    "\n",
    "# Read the contents of the reference sequences file\n",
    "with open(vamp_ref_seqs, 'r') as file1:\n",
    "    content1 = file1.read()\n",
    "\n",
    "# Read the contents of the denoised sequences file\n",
    "with open(denoised_sequences, 'r') as file2:\n",
    "    content2 = file2.read()\n",
    "\n",
    "# Write the combined contents to the new file\n",
    "with open(file_combined_path, 'w') as file_combined:\n",
    "    file_combined.write(content1 + '\\n' + content2)\n",
    "\n",
    "print(f\"Combined files {denoised_sequences} and {vamp_ref_seqs} into {file_combined_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61a425d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating alignment\n",
    "\n",
    "# paths\n",
    "input_sequences = f'{vamp_spec_dir}/reference_and_mock_seqs.fasta'\n",
    "\n",
    "\n",
    "\n",
    "#     creating an alingnment using clustalw:\n",
    "#     see more here: https://manpages.ubuntu.com/manpages/impish/man1/clustalw.1.html\n",
    "cmd = ClustalwCommandline(clustalw_path,\n",
    "                          infile=input_sequences,\n",
    "                          output='fasta',\n",
    "                          outfile=f'{vamp_spec_dir}/aligned_reference_and_mock_seqs.fasta')\n",
    "# print clustal command in a linux style\n",
    "print(cmd)\n",
    "# run clustalw alignment\n",
    "stdout, stderr = cmd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dae13065",
   "metadata": {},
   "outputs": [],
   "source": [
    "# if you want to check the resulting alignments using biopython\n",
    "\n",
    "align = AlignIO.read(f'{vamp_spec_dir}/aligned_reference_and_mock_seqs.fasta', 'fasta')\n",
    "print(align)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4465f9f",
   "metadata": {},
   "source": [
    "# Creating a maximum likelihood phylogenetic tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0aee2d2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to your alignment file\n",
    "alignment_path = f'{vamp_spec_dir}/aligned_reference_and_mock_seqs.fasta'\n",
    "\n",
    "# Output prefix for the tree\n",
    "output_prefix = f'{vamp_spec_dir}/raxml_tree'\n",
    "\n",
    "# The command to run RAxML-NG\n",
    "cmd = f\"raxml-ng --all --msa {alignment_path} --model GTR+G --prefix {output_prefix} --threads 2\"\n",
    "\n",
    "# Run the command\n",
    "process = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "stdout, stderr = process.communicate()\n",
    "\n",
    "# Check if the command was successful\n",
    "if process.returncode != 0:\n",
    "    print(f\"RAxML-NG encountered an error:\\n{stderr.decode()}\")\n",
    "else:\n",
    "    print(f\"RAxML-NG ran successfully. Tree(s) saved to {output_prefix}.bestTree and other files.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e98d6a52",
   "metadata": {},
   "source": [
    "# Tree visualizing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "5a497728",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'centroid=seq12_300_seqs=5': 'Vampyrella_lateritia_strain_VL09_seq12_300',\n",
       " 'centroid=seq173_22_seqs=10': 'Leptophrys_vorax_strain_LV02_seq173_22',\n",
       " 'centroid=seq199_18_seqs=1': 'Leptophrys_vorax_strain_LV02_seq199_18',\n",
       " 'centroid=seq179_25_seqs=3': 'Pseudovampyrella_closterii_strain_VC01_seq179_25',\n",
       " 'centroid=seq18_128_seqs=2': 'Placopus_flabellus_HR01_seq18_128',\n",
       " 'centroid=seq254_8_seqs=21': 'Leptophrys_vorax_strain_LV02_seq254_8',\n",
       " 'centroid=seq308_6_seqs=1': 'Leptophrys_vorax_strain_LV04_seq308_6'}"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Creating name_mapping dictionary to rename the new sequences in the tree\n",
    "\n",
    "name_mapping = {}\n",
    "\n",
    "with open(f'{tax_assign_dir}/blast6_Mock_18S.tab') as f:\n",
    "    content = f.read\n",
    "    for line in f:\n",
    "        if 'Vampyrellida' in line:\n",
    "            old_name = line.strip().split('\\t')[0].replace(';', '_')\n",
    "            new_name_second_part = line.strip().split(';')[0].strip('centroid=')\n",
    "            new_name_first_part = line.split(',s:')[1].split('\\t')[0]\n",
    "            name_mapping[old_name] = new_name_first_part + \"_\" + new_name_second_part\n",
    "            \n",
    "name_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "3be216e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from ete3 import Tree, TreeStyle, NodeStyle, TextFace\n",
    "\n",
    "os.environ['QT_QPA_PLATFORM'] = 'offscreen'\n",
    "line_width = 1\n",
    "tree_width = 200\n",
    "\n",
    "\n",
    "# Create a Tree object from the Newick file\n",
    "main_tree = Tree(f'{tree_file}/T2.raxml.bestTree')\n",
    "\n",
    "\n",
    "# Apply the renaming directly to the leaves of the tree\n",
    "for leaf in main_tree.iter_leaves():\n",
    "    if leaf.name in name_mapping:\n",
    "        leaf.name = name_mapping[leaf.name]\n",
    "        \n",
    "\n",
    "# Function to get color based on the label\n",
    "def get_color(name):\n",
    "    if 'Leptophrys_vorax_strain_LV02_' in name:\n",
    "        return 'red'\n",
    "    elif 'Leptophrys_vorax_strain_LV04_' in name:\n",
    "        return '#990000'\n",
    "    elif 'Vampyrella_lateritia_strain_VL09_' in name:\n",
    "        return 'green'\n",
    "    elif 'Vampyrella_lateritia_strain_NHR1' in name:\n",
    "        return '#33CC33'\n",
    "    elif 'Pseudovampyrella_closterii_strain_VC01_' in name:\n",
    "        return 'purple'\n",
    "    elif 'Placopus_flabellus_NHR1_' in name:\n",
    "        return 'orange'\n",
    "    elif 'Leptophrys_vorax_strain_LV01_' in name:\n",
    "        return '#FF6666'\n",
    "    return 'black'  # Default color if no match\n",
    "\n",
    "for node in main_tree.traverse():\n",
    "    # Set the color and thickness for all nodes in the subtree\n",
    "    nstyle = NodeStyle()\n",
    "    nstyle[\"size\"] = 0  # Set size to 0 to hide the default circle node\n",
    "    nstyle[\"vt_line_width\"] = line_width  # Set the thickness of vertical lines\n",
    "    nstyle[\"hz_line_width\"] = line_width  # Set the thickness of horizontal lines\n",
    "    node.set_style(nstyle)\n",
    "\n",
    "for leaf in main_tree.iter_leaves():\n",
    "    color = get_color(leaf.name)\n",
    "\n",
    "    # Apply the color to the text (leaf name)\n",
    "    name_face = TextFace(leaf.name, fgcolor=color)\n",
    "    leaf.add_face(name_face, column=0, position=\"branch-right\")\n",
    "\n",
    "    # Apply the color to the branch leading to the leaf\n",
    "    nstyle = NodeStyle()\n",
    "    nstyle[\"fgcolor\"] = color\n",
    "    nstyle[\"size\"] = 0  # Set size to 0 to hide the default circle node\n",
    "    nstyle[\"vt_line_color\"] = color\n",
    "    nstyle[\"hz_line_color\"] = color\n",
    "    nstyle[\"vt_line_width\"] = line_width  # Control the thickness of the vertical lines\n",
    "    nstyle[\"hz_line_width\"] = line_width  # Control the thickness of the horizontal lines\n",
    "    leaf.set_style(nstyle)\n",
    "\n",
    "\n",
    "ts = TreeStyle()\n",
    "ts.scale = tree_width  # Increase this value to spread out the tree more\n",
    "ts.show_leaf_name = False\n",
    "# main_tree.render(\"%%inline\", tree_style=ts)\n",
    "# main_tree.render(f\"{tree_file}/tree_mock.png\", dpi=600, w=10000, tree_style=ts)\n",
    "\n",
    "# save as Newick\n",
    "# main_tree.write(outfile=f\"{tree_file}/tree_mock.newick\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81abecc1",
   "metadata": {},
   "source": [
    "## Subtrees"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2c8de51",
   "metadata": {},
   "source": [
    "### Subtrees tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "aef4f292",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to get color based on the label\n",
    "def get_color(name, subtaxa):\n",
    "    \n",
    "    # color code for Leptophrys_vorax\n",
    "    if subtaxa == 'Leptophrys_vorax':\n",
    "        if 'Leptophrys_vorax_strain_LV02_' in name:\n",
    "            return 'red'\n",
    "        elif 'Leptophrys_vorax_strain_LV04_' in name:\n",
    "            return '#990000'\n",
    "        elif 'Leptophrys_vorax_strain_LV01_' in name:\n",
    "            return '#FF6666'\n",
    "        return 'black'  # Default color if no match\n",
    "    \n",
    "    # color code for Vampyrella_lateritia\n",
    "    elif subtaxa == 'Vampyrella_lateritia':\n",
    "        if 'Vampyrella_lateritia_strain_VL09_' in name:\n",
    "            return 'green'\n",
    "#         elif 'Vampyrella_lateritia_strain_NHR1' in name:\n",
    "#             return '#33CC33'\n",
    "    \n",
    "    # color code for Pseudovampyrella_closterii\n",
    "    elif subtaxa == 'Pseudovampyrella_closterii':\n",
    "        if 'Pseudovampyrella_closterii_strain_VC01_' in name:\n",
    "            return 'purple'\n",
    "    \n",
    "    # color code for Placopus_flabellus\n",
    "    elif subtaxa == 'Placopus_flabellus':\n",
    "        if 'Placopus_flabellus' in name:\n",
    "            return 'orange'\n",
    "    else:\n",
    "        print('Subtaxa name is probably incorrect. Choose among: Leptophrys_vorax, Vampyrella_lateritia, or Pseudovampyrella_closterii')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "190c09ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'nodes': [],\n",
       " 'faces': [[291.0749442908125,\n",
       "   8.206594517438232,\n",
       "   4624.156849498198,\n",
       "   147.71870131388818,\n",
       "   3,\n",
       "   'KF141791|Vampyrellida|Leptophryidae|Vernalophrys|Vernalophrys_algivore'],\n",
       "  [301.8253593246155,\n",
       "   147.71870131388818,\n",
       "   4208.164349625214,\n",
       "   287.2308081103381,\n",
       "   4,\n",
       "   'JF720748|Vampyrellida|Leptophryidae|Vernalophrys|Vernalophrys_X'],\n",
       "  [605.3515571892315,\n",
       "   287.2308081103381,\n",
       "   4963.053245948932,\n",
       "   426.74291490678803,\n",
       "   8,\n",
       "   'DQ409121|Vampyrellida|Leptophryidae|Leptophryidae_X|Leptophryidae_XX'],\n",
       "  [726.5086627751365,\n",
       "   426.74291490678803,\n",
       "   5084.210351534837,\n",
       "   566.255021703238,\n",
       "   9,\n",
       "   'DQ409086|Vampyrellida|Leptophryidae|Leptophryidae_X|Leptophryidae_XX'],\n",
       "  [233.74019572224998,\n",
       "   566.255021703238,\n",
       "   4673.507829656333,\n",
       "   705.7671284996879,\n",
       "   10,\n",
       "   'MF621964|Vampyrellida|Leptophryidae|Planctomyxa|Planctomyxa_polycarya'],\n",
       "  [238.67834582242796,\n",
       "   705.7671284996879,\n",
       "   4571.760251029814,\n",
       "   845.2792352961379,\n",
       "   12,\n",
       "   'EU567260|Vampyrellida|Leptophryidae|Leptophryidae_X|Leptophryidae_XX'],\n",
       "  [1114.5030915725365,\n",
       "   845.2792352961379,\n",
       "   5086.49483801264,\n",
       "   984.7913420925878,\n",
       "   16,\n",
       "   'AY620291|Vampyrellida|Vampyrellida_clade_A|clade_A_X|clade_A_XX'],\n",
       "  [1491.0214702826152,\n",
       "   984.7913420925878,\n",
       "   5479.426405757596,\n",
       "   1124.3034488890378,\n",
       "   18,\n",
       "   'EU567265|Vampyrellida|Vampyrellida_clade_A|clade_A_X|clade_A_XX'],\n",
       "  [1404.2877252883527,\n",
       "   1124.3034488890378,\n",
       "   5376.279471728456,\n",
       "   1263.8155556854877,\n",
       "   19,\n",
       "   'EF024716|Vampyrellida|Vampyrellida_clade_A|clade_A_X|clade_A_XX'],\n",
       "  [1511.3991226601217,\n",
       "   1263.8155556854877,\n",
       "   4506.806121525076,\n",
       "   1403.3276624819375,\n",
       "   20,\n",
       "   'Pseudovampyrella_closterii_strain_VC01_seq222_17'],\n",
       "  [428.68456742677876,\n",
       "   1403.3276624819375,\n",
       "   4704.320311012098,\n",
       "   1542.8397692783876,\n",
       "   22,\n",
       "   'FJ410735|Vampyrellida|Leptophryidae|Leptophryidae_X|Leptophryidae_XX'],\n",
       "  [430.95008148825804,\n",
       "   1542.8397692783876,\n",
       "   4747.618797660768,\n",
       "   1682.3518760748375,\n",
       "   24,\n",
       "   'AY642743|Vampyrellida|Leptophryidae|Leptophryidae_X|Leptophryidae_XX'],\n",
       "  [430.9524763234183,\n",
       "   1682.3518760748375,\n",
       "   4764.034381530805,\n",
       "   1821.8639828712874,\n",
       "   25,\n",
       "   'EU567261|Vampyrellida|Leptophryidae|Leptophryidae_X|Leptophryidae_XX'],\n",
       "  [299.18625097815755,\n",
       "   1821.8639828712874,\n",
       "   5288.795717580602,\n",
       "   1961.3760896677375,\n",
       "   27,\n",
       "   'XXXXXXXX|Vampyrellida|Leptophryidae|Leptophryidae_X|Leptophryidae_X_strain_VI01'],\n",
       "  [276.483213660168,\n",
       "   1961.3760896677375,\n",
       "   5815.934512930974,\n",
       "   2100.8881964641873,\n",
       "   31,\n",
       "   'OQ591879|Vampyrellida|Leptophryidae|Pseudovampyrella|Pseudovampyrella_minor_strainVC02'],\n",
       "  [280.00362134555274,\n",
       "   2100.8881964641873,\n",
       "   5991.793405482562,\n",
       "   2240.4003032606374,\n",
       "   32,\n",
       "   'OQ591878|Vampyrellida|Leptophryidae|Pseudovampyrella|Pseudovampyrella_closterii_strain_VC01'],\n",
       "  [263.5606831361312,\n",
       "   2240.4003032606374,\n",
       "   3258.967682001086,\n",
       "   2379.912410057087,\n",
       "   33,\n",
       "   'Pseudovampyrella_closterii_strain_VC01_seq179_25'],\n",
       "  [195.89701052194792,\n",
       "   2379.912410057087,\n",
       "   5210.126260676707,\n",
       "   2519.424516853537,\n",
       "   35,\n",
       "   'XXXXXXXX|Vampyrellida|Leptophryidae|Strigomyxa|Strigomyxa_ruptor_strain_Strigo01'],\n",
       "  [195.89701052194792,\n",
       "   2519.424516853537,\n",
       "   4282.781080206187,\n",
       "   2658.9366236499873,\n",
       "   36,\n",
       "   'XXXXXXXX|Vampyrellida|Leptophryidae|Strigomyxa|Strigomyxa_ruptor']],\n",
       " 'node_areas': {0: [8.206594517438232,\n",
       "   8.206594517438232,\n",
       "   5991.793405482562,\n",
       "   2658.9366236499873],\n",
       "  1: [34.212109521240166,\n",
       "   8.206594517438232,\n",
       "   5479.426405757596,\n",
       "   1821.8639828712874],\n",
       "  2: [50.85860871927196,\n",
       "   8.206594517438232,\n",
       "   4624.156849498198,\n",
       "   287.2308081103381],\n",
       "  3: [291.0725494556522,\n",
       "   8.206594517438232,\n",
       "   4624.156849498198,\n",
       "   147.71870131388818],\n",
       "  4: [291.0725494556522,\n",
       "   147.71870131388818,\n",
       "   4208.164349625214,\n",
       "   287.2308081103381],\n",
       "  5: [50.85860871927196,\n",
       "   287.2308081103381,\n",
       "   5479.426405757596,\n",
       "   1821.8639828712874],\n",
       "  6: [62.7609394650953,\n",
       "   287.2308081103381,\n",
       "   5084.210351534837,\n",
       "   705.7671284996879],\n",
       "  7: [107.45095838819577,\n",
       "   287.2308081103381,\n",
       "   5084.210351534837,\n",
       "   566.255021703238],\n",
       "  8: [605.3491623540713,\n",
       "   287.2308081103381,\n",
       "   4963.053245948932,\n",
       "   426.74291490678803],\n",
       "  9: [605.3491623540713,\n",
       "   426.74291490678803,\n",
       "   5084.210351534837,\n",
       "   566.255021703238],\n",
       "  10: [107.45095838819577,\n",
       "   566.255021703238,\n",
       "   4673.507829656333,\n",
       "   705.7671284996879],\n",
       "  11: [62.7609394650953,\n",
       "   705.7671284996879,\n",
       "   5479.426405757596,\n",
       "   1821.8639828712874],\n",
       "  12: [155.91763235881405,\n",
       "   705.7671284996879,\n",
       "   4571.760251029814,\n",
       "   845.2792352961379],\n",
       "  13: [155.91763235881405,\n",
       "   845.2792352961379,\n",
       "   5479.426405757596,\n",
       "   1821.8639828712874],\n",
       "  14: [329.35398949025705,\n",
       "   845.2792352961379,\n",
       "   5479.426405757596,\n",
       "   1403.3276624819375],\n",
       "  15: [449.22267376001474,\n",
       "   845.2792352961379,\n",
       "   5479.426405757596,\n",
       "   1263.8155556854877],\n",
       "  16: [1065.4401036470467,\n",
       "   845.2792352961379,\n",
       "   5086.49483801264,\n",
       "   984.7913420925878],\n",
       "  17: [1065.4401036470467,\n",
       "   984.7913420925878,\n",
       "   5479.426405757595,\n",
       "   1263.8155556854877],\n",
       "  18: [1263.8689605095587,\n",
       "   984.7913420925878,\n",
       "   5479.426405757596,\n",
       "   1124.3034488890378],\n",
       "  19: [1263.8689605095587,\n",
       "   1124.3034488890378,\n",
       "   5376.279471728456,\n",
       "   1263.8155556854877],\n",
       "  20: [449.22267376001474,\n",
       "   1263.8155556854877,\n",
       "   4506.806121525076,\n",
       "   1403.3276624819375],\n",
       "  21: [329.35398949025705,\n",
       "   1403.3276624819375,\n",
       "   4764.034381530804,\n",
       "   1821.8639828712874],\n",
       "  22: [422.1778002967184,\n",
       "   1403.3276624819375,\n",
       "   4704.320311012098,\n",
       "   1542.8397692783876],\n",
       "  23: [422.1778002967184,\n",
       "   1542.8397692783876,\n",
       "   4764.034381530805,\n",
       "   1821.8639828712874],\n",
       "  24: [422.1801951318785,\n",
       "   1542.8397692783876,\n",
       "   4747.618797660768,\n",
       "   1682.3518760748375],\n",
       "  25: [422.1801951318785,\n",
       "   1682.3518760748375,\n",
       "   4764.034381530805,\n",
       "   1821.8639828712874],\n",
       "  26: [34.212109521240166,\n",
       "   1821.8639828712874,\n",
       "   5991.793405482562,\n",
       "   2658.9366236499873],\n",
       "  27: [60.921706062118986,\n",
       "   1821.8639828712874,\n",
       "   5288.795717580602,\n",
       "   1961.3760896677375],\n",
       "  28: [60.921706062118986,\n",
       "   1961.3760896677375,\n",
       "   5991.793405482562,\n",
       "   2658.9366236499873],\n",
       "  29: [93.95366742572891,\n",
       "   1961.3760896677375,\n",
       "   5991.793405482562,\n",
       "   2379.912410057087],\n",
       "  30: [248.98811118676824,\n",
       "   1961.3760896677375,\n",
       "   5991.793405482562,\n",
       "   2240.4003032606374],\n",
       "  31: [263.457705224246,\n",
       "   1961.3760896677375,\n",
       "   5815.934512930974,\n",
       "   2100.8881964641873],\n",
       "  32: [263.457705224246,\n",
       "   2100.8881964641873,\n",
       "   5991.793405482562,\n",
       "   2240.4003032606374],\n",
       "  33: [248.98811118676824,\n",
       "   2240.4003032606374,\n",
       "   3258.967682001086,\n",
       "   2379.912410057087],\n",
       "  34: [93.95366742572891,\n",
       "   2379.912410057087,\n",
       "   5210.126260676707,\n",
       "   2658.9366236499873],\n",
       "  35: [195.8946156867877,\n",
       "   2379.912410057087,\n",
       "   5210.126260676707,\n",
       "   2519.424516853537],\n",
       "  36: [195.8946156867877,\n",
       "   2519.424516853537,\n",
       "   4282.781080206187,\n",
       "   2658.9366236499873]}}"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Subtaxa\n",
    "subtaxa = 'Pseudovampyrella_closterii'\n",
    "\n",
    "# load the tree\n",
    "main_tree = Tree(f\"{tree_file}/tree_mock.newick\")\n",
    "\n",
    "# Collect leaves that match the condition for new sequences\n",
    "matching_leaves = [leaf for leaf in main_tree.iter_leaves() if f'{subtaxa}_strain_' in leaf.name and 'seq' in leaf.name]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Find the common ancestor of these leaves\n",
    "common_ancestor = main_tree.get_common_ancestor(matching_leaves)\n",
    "\n",
    "# Use this common ancestor as the root of a new subtree\n",
    "subtree = common_ancestor.copy(method=\"newick\")\n",
    "\n",
    "\n",
    "for node in subtree.traverse():\n",
    "    # Set the color and thickness for all nodes in the subtree\n",
    "    nstyle = NodeStyle()\n",
    "    nstyle[\"size\"] = 0  # Set size to 0 to hide the default circle node\n",
    "    nstyle[\"vt_line_width\"] = 2  # Set the thickness of vertical lines\n",
    "    nstyle[\"hz_line_width\"] = 2  # Set the thickness of horizontal lines\n",
    "    node.set_style(nstyle)\n",
    "\n",
    "for leaf in subtree.iter_leaves():\n",
    "    color = get_color(name = leaf.name, subtaxa = subtaxa)\n",
    "\n",
    "    # Apply the color to the text (leaf name)\n",
    "    name_face = TextFace(leaf.name, fgcolor=color)\n",
    "    leaf.add_face(name_face, column=0, position=\"branch-right\")\n",
    "\n",
    "    # Apply the color to the branch leading to the leaf\n",
    "    nstyle = NodeStyle()\n",
    "    nstyle[\"fgcolor\"] = color\n",
    "    nstyle[\"size\"] = 0  # Set size to 0 to hide the default circle node\n",
    "    nstyle[\"vt_line_color\"] = color\n",
    "    nstyle[\"hz_line_color\"] = color\n",
    "    nstyle[\"vt_line_width\"] = 2  # Control the thickness of the vertical lines\n",
    "    nstyle[\"hz_line_width\"] = 2  # Control the thickness of the horizontal lines\n",
    "    leaf.set_style(nstyle)\n",
    "\n",
    "\n",
    "ts = TreeStyle()\n",
    "ts.show_leaf_name = False\n",
    "# subtree.render(\"%%inline\", tree_style=ts)\n",
    "subtree.render(f\"{tree_file}/subtree_mock_{subtaxa}.png\", dpi=600, w=6000, tree_style=ts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c343a09",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d468f06",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
